{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "install notebook extensions: pip install jupyter_contrib_nbextensions && jupyter contrib nbextension install --user"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "list of laTex mathematical symbols\n",
    "https://oeis.org/wiki/List_of_LaTeX_mathematical_symbols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "        <script type=\"text/javascript\">\n",
       "        window.PlotlyConfig = {MathJaxConfig: 'local'};\n",
       "        if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}\n",
       "        if (typeof require !== 'undefined') {\n",
       "        require.undef(\"plotly\");\n",
       "        requirejs.config({\n",
       "            paths: {\n",
       "                'plotly': ['https://cdn.plot.ly/plotly-latest.min']\n",
       "            }\n",
       "        });\n",
       "        require(['plotly'], function(Plotly) {\n",
       "            window._Plotly = Plotly;\n",
       "        });\n",
       "        }\n",
       "        </script>\n",
       "        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "        <script type=\"text/javascript\">\n",
       "        window.PlotlyConfig = {MathJaxConfig: 'local'};\n",
       "        if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}\n",
       "        if (typeof require !== 'undefined') {\n",
       "        require.undef(\"plotly\");\n",
       "        requirejs.config({\n",
       "            paths: {\n",
       "                'plotly': ['https://cdn.plot.ly/plotly-latest.min']\n",
       "            }\n",
       "        });\n",
       "        require(['plotly'], function(Plotly) {\n",
       "            window._Plotly = Plotly;\n",
       "        });\n",
       "        }\n",
       "        </script>\n",
       "        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "        <script type=\"text/javascript\">\n",
       "        window.PlotlyConfig = {MathJaxConfig: 'local'};\n",
       "        if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}\n",
       "        if (typeof require !== 'undefined') {\n",
       "        require.undef(\"plotly\");\n",
       "        requirejs.config({\n",
       "            paths: {\n",
       "                'plotly': ['https://cdn.plot.ly/plotly-latest.min']\n",
       "            }\n",
       "        });\n",
       "        require(['plotly'], function(Plotly) {\n",
       "            window._Plotly = Plotly;\n",
       "        });\n",
       "        }\n",
       "        </script>\n",
       "        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import binom\n",
    "import  statsmodels.stats.multitest as multitest\n",
    "import itertools\n",
    "\n",
    "import seaborn as sns\n",
    "# plotly.offline.init_notebook_mode(connected=True)\n",
    "# Standard plotly imports\n",
    "import plotly.plotly as py\n",
    "import pandas as pd\n",
    "import plotly\n",
    "import plotly.graph_objs as go\n",
    "from plotly.offline import iplot\n",
    "\n",
    "from tqdm import tnrange, tqdm_notebook\n",
    "plotly.io.orca.config.port = 8999\n",
    "\n",
    "from plotly import tools\n",
    "# cufflinks binds plotly directly to python dataframe\n",
    "import cufflinks as cf\n",
    "cf.go_offline()\n",
    "# # Set the global theme\n",
    "cf.set_config_file(world_readable=True, theme='pearl', offline=True)\n",
    "# plotly.offline.init_notebook_mode(connected=True)\n",
    "\n",
    "\n",
    "%matplotlib inline\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# read text in files as a list of texts "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['The special effects of this movie are, especially for its time, laughable and used in such an over-emphasized way that you can\\'t deny their terrible existance.<br /><br />The acting redefines the term \"terrible overacting\" at the hands of Meg Foster and Richard Joseph Paul, where julie Newman and Andrew Divoff just redefine \"bad\".<br /><br />***spoilers***<br /><br />The charm in this movie can be found in two things: First is the excellent casting of Carel \"Lurch\" Struycken as the mysterious psychic Gaunt, who can sense where and when people will die and is always there.<br /><br />The second are original finds, the combination SF-Western is obviously original, if terrible, but other finds are more original, like the gunman Zack Stone being able to sense the pain of the people he shoots (though his acting falls short here).<br /><br />Overal...don\\'t see this movie, except if you love that ol\\' hunk-o-brutal Carel Struycken, as any self-respecting Dutchman should.',\n",
       " \"You get 5 writers together, have each write a different story with a different genre, and then you try to make one movie out of it. It's action, it's adventure, it's sci-fi, it's western, it's a mess. Sorry, but this movie absolutely stinks. 4.5 is giving it an awefully high rating. That said, it's movies like this that make me think I could write movies, and I can barely write.\"]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = '/projects/trans_scratch/validations/workspace/szong/development/fastai/courses/ml1/data/aclImdb/train/neg/'\n",
    "fs = ['12265_1.txt', '12264_1.txt']\n",
    "txts = []\n",
    "for f in fs:\n",
    "    fname = f'{path}{f}'\n",
    "    with open(fname) as fh:\n",
    "        txts.append(fh.read())\n",
    "\n",
    "txts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# useful numpy functions: sign, clip, nonzero, argmax, argsort "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-1.,  1.,  1., -1.,  0.,  0.])"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array([-2. ,  3. ,  0.1, -0.2,  0. ,  0. ])"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(array([0, 1, 2, 3]),)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array([-5.3,  5.4,  0.1, -0.2])"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "5.4"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array([0, 3, 4, 5, 2, 1])"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array([-5.3, -0.2,  0. ,  0. ,  0.1,  5.4])"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array([-5.3, -0.2,  0. ,  0. ,  0.1,  5.4])"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array([100., 100., 100., 100., 100., 100.])"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = np.array([-5.3, 5.4, 0.1, -0.2, 0, 0])\n",
    "np.sign(a)\n",
    "\n",
    "\n",
    "np.clip(a, -2, 3)\n",
    "\n",
    "np.nonzero(a)\n",
    "a[np.nonzero(a)]\n",
    "\n",
    "np.argmax(a)\n",
    "a[np.argmax(a)]\n",
    "\n",
    "np.argsort(a)\n",
    "a[np.argsort(a)]\n",
    "\n",
    "# ndarray inplace operation\n",
    "a.sort()\n",
    "a\n",
    "a.fill(100)\n",
    "a\n",
    "\n",
    "# generate synthetic dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# generate synthetic dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.linspace()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.        , 0.02040816, 0.04081633, 0.06122449, 0.08163265,\n",
       "       0.10204082, 0.12244898, 0.14285714, 0.16326531, 0.18367347,\n",
       "       0.20408163, 0.2244898 , 0.24489796, 0.26530612, 0.28571429,\n",
       "       0.30612245, 0.32653061, 0.34693878, 0.36734694, 0.3877551 ,\n",
       "       0.40816327, 0.42857143, 0.44897959, 0.46938776, 0.48979592,\n",
       "       0.51020408, 0.53061224, 0.55102041, 0.57142857, 0.59183673,\n",
       "       0.6122449 , 0.63265306, 0.65306122, 0.67346939, 0.69387755,\n",
       "       0.71428571, 0.73469388, 0.75510204, 0.7755102 , 0.79591837,\n",
       "       0.81632653, 0.83673469, 0.85714286, 0.87755102, 0.89795918,\n",
       "       0.91836735, 0.93877551, 0.95918367, 0.97959184, 1.        ])"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array([-0.03135407, -0.11814905,  0.14586559, -0.04239754,  0.16058562,\n",
       "       -0.05277152,  0.10170596, -0.00888891,  0.19051268,  0.0203645 ,\n",
       "        0.3097905 ,  0.04893543,  0.33457591,  0.10317423,  0.35499112,\n",
       "        0.48008651,  0.37169467,  0.30716088,  0.31143893,  0.35979426,\n",
       "        0.60311788,  0.5898593 ,  0.5409102 ,  0.54906796,  0.53380142,\n",
       "        0.55232376,  0.55209233,  0.66242741,  0.58648964,  0.6274728 ,\n",
       "        0.76066367,  0.54772801,  0.53610087,  0.85922766,  0.7635116 ,\n",
       "        0.73061079,  0.85773251,  0.72330783,  0.58765447,  0.98966587,\n",
       "        0.64534461,  0.89153296,  0.69251614,  0.68548496,  0.84034004,\n",
       "        1.00092404,  0.9832276 ,  1.07908551,  0.92975307,  0.93793445])"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x7f34d2da4160>"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x7f34d2d74320>"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXUAAAD4CAYAAAATpHZ6AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAe30lEQVR4nO3df5BdZZ3n8feHTiON49hoqF1paBNnIQyIEm1/1KZqBcRJYHZCBl0lrjuyG02x/phydFKGckspLIuM1BQ6NYwOOhTjVA0/BCqbRSycNVDWMsah2SBINE4GdUxjLRkkVE0RIYnf/eOeG27f3HvPubfPufeccz+vqi667z1970Pz6S+nn/Oc76OIwMzM6uGEUQ/AzMzy46JuZlYjLupmZjXiom5mViMu6mZmNbJsVG+8fPnyWLFixaje3mru4Ycf/peIOHUU7+1sW5HSsj2yor5ixQrm5+dH9fZWc5J+Nqr3dratSGnZ9vSLmVmNuKibmdWIi7qZWY24qJuZ1YiLuo0tSTdLekrSD1KOe5Oko5LeNayxmQ3KRd3G2S3Aul4HSJoA/gS4bxgDMlsqF3UbWxHxHeCXKYd9FLgLeKr4EZkt3cjWqZt1s333Atfft5cnDx7itOkptqxdxYbVM0Mfh6QZ4PeBi4A3pRy7GdgMMDs7W/zgrJKGkW2fqVupbN+9wNV3P8bCwUMEsHDwEFff/Rjbdy+MYjhfAD4ZEUfTDoyImyJiLiLmTj11JDeyWskNK9su6lYq19+3l0OHF9fQQ4ePcv19e0cxnDngNkk/Bd4F/IWkDaMYiFXfsLLt6RcrlScPHurr8SJFxMrm55JuAe6JiO1DH4jVwrCy7TN1K5XTpqf6enwpJN0KfBdYJWm/pE2SrpJ0Ve5vZmNvWNl2UbdS2bJ2FVOTE4sem5qcYMvaVbm/V0RsjIhXRcRkRJweEX8VEV+OiC93OPbKiLgz90HY2BhWtj39YqXSXAlQhtUvZnkaVrZd1K10NqyecRG3WhpGtj39YmZWIy7qZmY14qJuZlYjLupmZjXiom5mViOpq18k3Qz8R+CpiHhth+cFfBG4FHgOuDIi/m/eA7V6KkvzLrM8jTLXWc7Ub6F3z+lLgDOTj83Al5Y+LBsHJWveZZaLUec6tahn6Dl9GfC1aNgFTEt6VV4DtHLbvnuBNdt2snLrN1izbWdfwS1Z8y6zY6qc6zzm1GeAn7d8vT957DiSNkualzR/4MCBHN7aRmmpZyTdGhktHDzU9y+SWV6qnus8iro6PBadDnTP6XpZ6hlJr0ZGnoqxUal6rvMo6vuBM1q+Ph14MofXtZJbaivRTg2OWnkqxkah6rnOo6jvAP5ADW8Fno2IX+TwulZyS20lumH1DNddfh4zPY4fRR91G29Vz3VqUc/Qc/pe4AlgH/AV4EOFjdZKJY9WohtWz/Dg1ou6/gIU0UfdrJeq5zp1nXpEbEx5PoAP5zYiq4w8W4luWbuKq+9+bNFcZlF91M16qXqu3XrXliSvVqLuo25lUuVcu6hbaaT9IvnuU6uiYefaRd0qobl2uPlnbHNpGODCbpVVRK7d0MtKrXln38duf8R3n1qtbN+9wCfu+H7uufaZupVW+1lMJ17yaFXUzPbR6Hif5pJy7TN1K61Od/a185JHq6K0bC8l1y7qVlppZyte8mhV1SvbS821i7qVVq+zlZnpKa67/LwlXSSVdLOkpyT9oMvz/1nSo8nH30t6/cBvZtaiW7YnpCXn2kXdSqvbnX1feM/5PLj1ojxWvdxC770CfgK8LSJeB3wWuGmpb2gG3bP9p+9+/ZJz7Qullru81t0WfeNGRHxH0ooez/99y5e7aDSrszFWhWy7qFuu8l53m9edfTnYBHyz25OSNtPY+YvZ2dlhjcmGqCrZ9vSL5WrUu74UQdKFNIr6J7sd470C6q8q2faZuuVqqb2oy0bS64CvApdExNOjHo+NTlWy7TN1y9VSe1GXiaRZ4G7gv0TEj0c9HhutqmTbZ+pjLO2izyAXharUQjfZK+ACYLmk/cBngEmAiPgy8GnglcBfSAI4EhFzoxmt9WOcs+2iPqbSLvoMelGoSi10M+wV8AHgA0MajuVk3LPtoj6mel302bB6JvX5Xkq0YsXG0Lhn20V9TKVd9CnyopD7oluRxj3bvlA6ptIu+hR1Uaj5p+/CwUMEL/7pu333wpJe16xp3LPtoj6m0jbXzWPz3U6qstbXqmvcs+3plzGVdtGnqItCVVnra9U17tl2UR9jaRd9irgodNr0FAsdQl62tb5WbeOcbU+/2FAV9aev2aiVJduZirqkdZL2StonaWuH52cl3S9pd9J7+tL8h2p1sGH1DNddfh4z01OIfPqim5VBWbKdOv0iaQK4EXgHsB94SNKOiNjTctj/AO6IiC9JOge4F1hRwHitBqqw1tdsEGXIdpYz9TcD+yLiiYh4AbgNuKztmAB+M/n85cCT+Q3RzMyyynKhdAb4ecvX+4G3tB1zDfAtSR8FXgpcnMvozMysL1nO1NXhsWj7eiNwS0ScDlwK/I2k415b0mZJ85LmDxw40P9ozcyspyxFfT9wRsvXp3P89Mom4A6AiPgucBKwvP2FvJGAmVmxsky/PAScKWklsABcAby37Zh/Bt4O3CLpt2kUdZ+KV0gZelaYFWHcsp1a1CPiiKSPAPcBE8DNEfG4pGuB+YjYAXwC+IqkP6IxNXNlRLRP0diQZQ1z3nsvmhXN2e4u0x2lEXEvjWWKrY99uuXzPcCafIdmS9FPmJfSitRs2Jzt3nxHaU3101yoLD0rzLJwtntzUa+pfsJclb0XzcDZTuOiXlP9hLksPSvMsnC2e3NRr6l+wtzes2J6apKTJk/gj25/hDXbdnoDCyuVQbMNMCEdm6qpa67derem+u0Z3exZ0e0i1PzPfsn9PzowNsvCrLwGyTbQ9eJqP69VBRrVysO5ubmYn58fyXtbd2u27ezYE1osvo14anKi1N0VJT0cEXOjeG9nu3y65Xp6apLnj/x60YXXqmfb0y+2SLeLUO3/6/cWdFYl3XJ98NDhUmxBlycXdVukn1UBVV8WJulmSU9J+kGX5yXpz5J9BB6V9IZhj9Hy0e9qlypn20XdFul0EapTRzeoxbKwW4B1PZ6/BDgz+dgMfGkIY7ICdLu4esrJkx2Pr3K2faHUFul0EerCs0/lrocXjpt3zLIsrMx9NyLiO5JW9DjkMuBrScuLXZKmJb0qIn4xlAFabrpdXIXFF1AhW7bLnGtfKLVMBglx+0qaplNOnuQzv3duob8EWS+UJkX9noh4bYfn7gG2RcT/Sb7+NvDJiOgZXGe7WvrNdqdcNxcSzAyhwKdl22fqlskg23R1up0b4JnnDlelqVKWvQQaB0qbaUzRMDs7W+SYLGf9ZrtTrpuhKEPDMM+pW2F6XWyqyAqDLHsJAN4rYJykXUQddbZd1K0waRebKrDCYAfwB8kqmLcCz3o+3bJcRB1ltl3UK2z77gXWbNvJyq3fKOXt/J1WHLQa9QoDSbcC3wVWSdovaZOkqyRdlRxyL/AEsA/4CvChEQ117JQ522m5htFm23PqFVWF5v/NcVyz43EOHjq86LkyNFWKiI0pzwfw4SENxxJlz3brSpqFg4c63m09ymy7qFdU2Zr/d1tB0NpTpqxLwKxcqpTtXs+Piot6RZWp+X+WM6tBVs/YeHK2l8Zz6hVVpub//exEY5bG2V4aF/WK6nbb84Vnnzr0C0xlOrOy6nO2l8bTLxWV5Xb+YV1gOm16qmNb01GvbrFqcraXxkW9wtrn8tZs2zmSC0xb1q4aqH+GWTfO9uBc1GtkVH8q9rsTjVm/nO3sXNSHrMjlT6P8U7FsKwBs+Jztcsh0oVTSOkl7k80CtnY55t2S9kh6XNLf5jvMemguj1o4eIjgxXnBvC74jOPO6VYOznZ5pBZ1SRPAjTQ2DDgH2CjpnLZjzgSuBtZExLnAxwoYa+V1Wx71sdsfyeVqfuvO6aLRBrTMey1afXTL9ifu+H4uhd3Zzi7L9MubgX0R8QSApNtobB6wp+WYDwI3RsQzABHxVN4DrYNe839LvZrf/qfvDe85v7DAl+0OOhu9btk+GpFrrovMWl1ynWX6ZQb4ecvX+5PHWp0FnCXpQUm7JHXcIkzSZknzkuYPHDgw2IgrLG3+b9CbGor+03dU72XV0SvbzvVwZSnqWTYKWEZjH8cLgI3AVyVNH/dNY95zOkt3t0Gu5g/zrrcq3mFnxUvLtnM9PFmKepaNAvYD/zMiDkfET4C9NIq8tWidF+xmkKv5w1zuVcU77Kx4zWxPqPM25c718GQp6g8BZ0paKelE4Aoamwe02g5cCCBpOY3pmCfyHGhdbFg9w4NbL+IL7zk/t6v5w+yVUaa+HFYuG1bP8Kfvfr1zPWKpRT0ijgAfAe4DfgjcERGPS7pW0vrksPuApyXtAe4HtkTE00UNug76vZrfa9OAYS738tIy68W5Hj019gEYPu+43lmnK/DAcbcqA5xy8iSf+b1zM/Urz/PKfhVWCaTtuF4kZ7uz9ty093MBjm04MdOSfed6sbRsu6iXQDNM3XZROWnyBJ557nDH752anEhdr9veEzrr91WZi3o59Mp2+9ft0jI6jrmG9Gy79e6ItS6lguNDfujw0a4Fvfl82hX6om96MuskLdtpp5Np2XauO3NRH7FOwexX2hX6LDc9jesvgBWn6Gw71525qI9YHkum0q7QF3XTk1kvRWfbue7MRX3ElrpkKssV+qJuejLrpd9st69wT8u2c92Zi/qIdQpm59s3GqanJvtualTUTU9mvWQpuk0zSb+ifrLtXHfmfuoj1q0JPxy/jHFqcoJr1p870JX9Zk/obisGqrge18oty7Z08GL+Bulb7lwfz0W9BHqFOe91s1XcyaVISfO5LwITwFcjYlvb87PAXwPTyTFbI+LeoQ+0ojple+7Vr3CuC+R16lZLWdapJ3sF/Bh4B43+RQ8BGyNiT8sxNwG7I+JLyT4C90bEil6v62xbkbxO3ay7Y3sFRMQLQHOvgFYB/Gby+cs5vpmdWal4+mVAVbml2HrqtFfAW9qOuQb4lqSPAi8FLu70QpI2A5sBZmdncx/oMDnb1eYz9QHUqaH+mMuyV8BG4JaIOB24FPgbScf93tRlrwBnu/pc1AdQp4b6Yy7LXgGbgDsAIuK7wEnA8qGMbgSc7epzUR9AnRrqj7ksewX8M/B2AEm/TaOo13YvRme7+lzUB1CnhvrjLONeAZ8APijp+8CtwJUxqiVjQ+BsV5+L+gDq1FB/3EXEvRFxVkT8VkR8Lnns0xGxI/l8T0SsiYjXR8T5EfGt0Y64WM529Xn1ywB8o4PVlbNdfS7qAxrklmazKnC2q83TL2ZmNeKibmZWIy7qZmY14qJuZlYjLupmZjWSqahLWidpr6R9krb2OO5dkkJSz5anZmZWjNSinvScvhG4BDgH2Jj0lW4/7mXAHwLfy3uQZmaWTZYz9Sw9pwE+C3we+FWO4zMzsz5kKeqdek4vujNB0mrgjIi4p9cLSdosaV7S/IEDte2JZGY2MlmKes+e00lv6RtoND7qqS49p83MyipLm4C0ntMvA14LPCAJ4N8COyStj4ix2KjRO8VYXTnb1ZOlqB/rOQ0s0Og5/d7mkxHxLC2bBkh6APjjcSroV9/92LGNBZo7xQAOv1Was11NqUU9Io5IavacngBubvacBuabLUrHRfuZy3MvHOm6U0wz+D7bsbLrlNFeuyA52+WVqUtjRNwL3Nv22Ke7HHvB0odVTp3OXLpp7hTjsx0ru24ZbS/oTc52ufmO0j50OnPpprlTjPd8tLLrltEJdVoj4WyXnYt6H7Lu09i6U4z3fLSy65bFoxE9d0FytsvJm2R00Wmu8LTpqY5TLtNTk7z0Jcs6zit2+x7v+Wij0p7t6ZMneea5w8cdN9Myt+5sV4eLegfd5grf+cYZ7np4YdGfnFOTE1yz/tyuc4hb1q46bn7Sez7aqHTK9uQJYnJCHD764n7azYz22gXJ2S4nT7900G2u8P4fHeC6y89jZnoK0TiTue7y83peFNqweqbv7zErSqdsH/518NITl/WdUWe7nHym3kGvucJ+9m9s/zP3hvec78DbSHXL9rOHDvPIZ34n02t4GWO5uaiTfY4xy1xh87UWDh5CvNhPwcu9bNj6uS6Ulm3nujrGfvqlOce4cPAQQSOk//qrI0xOLF7OlWWusPW1oKVBTsLLvcolyz4Bkt4taY+kxyX97bDHOKhOub767se48OxTe65oSXstcK7LbuyLep5zjFnWsXu5Vzlk2SdA0pnA1cCaiDgX+NjQBzqgPK8LOdfVMvbTL3nMMaa9Visv9yqNY/sEAEhq7hOwp+WYDwI3RsQzABHx1NBHOaC8rgv1eq1WznV5jP2ZercwDhLStO/xcq9SSd0nADgLOEvSg5J2SVrX7cXKtleAcz2+xr6ob1m7qu85xn5eqzkz7+VepdNzn4DEMuBM4AJgI/BVSdOdXqxsewU41+Nr7KdfmmHMY4lWnq9lhUvbJ6B5zK6IOAz8RNJeGkX+oeEMcXDO9fhSRPvJyXDMzc3F/PxYtFy3EZD0cETM9Xh+GfBj4O009gl4CHhvRDzecsw6YGNEvF/ScmA3cH5EPN3rvZ1tK1Jatsd++sXGU0QcAZr7BPwQuKO5T4Ck9clh9wFPS9oD3A9sSSvoZqM29tMvNr7S9gmIxp+xH08+zCrBZ+pmZjXiom5mViMu6mZmNeKibmZWIy7qZmY14qJuZlYjLupmZjWSqain9Z2W9PGk5/Sjkr4t6dX5D9XMzNKkFvUsfadp3D49FxGvA+4EPp/3QM3MLF2WO0pT+05HxP0tx+8C3pfnIPPivRWtrpxta8pS1Dv1nX5Lj+M3Ad/s9ISkzcBmgNnZ2YxDzEdzS67mDi7eW9Hqwtm2Vlnm1LP0nW4cKL0PmAOu7/T8KHtOd9vey3srWtU529Yqy5l6lr7TSLoY+BTwtoh4Pp/h5afX9l5mVeZsW6ssZ+oPAWdKWinpROAKYEfrAZJWA38JrC/rPo55bu9lVibOtrVKLeoZ+05fD/wG8HVJj0ja0eXlcrd99wJrtu1k5dZvsGbbTrbvXuh4XJ7be5kNg7Ntg8jUTz1D3+mLcx5XJv1cIPKWXFYlzrYNqlKbZLQv23ruhSNdLxB1CvSG1TMOupWSs215qUxR73Tm0o0vEFmVONuWp8r0fum0bKsbXyCyKnG2LU+VKepZz1B8gciqxtm2PFWmqHc7Q5memmRmegoBM9NTXHf5eZ5btEpxti1PlZlT37J21aJ5R2icuVyz/lwH3SrN2bY8Vaaoe9mW1ZWzbXmqTFGH/pZt5dm1zh3w6kvSOuCLwATw1YjY1uW4dwFfB94UEfN5j8PZtrxUqqhnlWfXOnfAq6+WvQLeQaPH0UOSdkTEnrbjXgb8IfC94Y9yMWfb0lTmQmk/8uxa5w54tXZsr4CIeAFo7hXQ7rM0Nn751TAH14mzbWlqWdTz7FrnDni11mmvgEWnqEmzujMi4p5eLyRps6R5SfMHDhzIf6QJZ9vS1LKo59m1zh3waq3nXgGSTgBuAD6R9kLD2ivA2bY0tSzqeXatcwe8WkvbK+BlwGuBByT9FHgrsEPS3NBG2MbZtjS1vFCa5xIxLzertWN7BQALNPYKeG/zyYh4Flje/FrSA8AfF7H6JStn29IoouPOdIWbm5uL+fmR/W5YzUl6OCJSz6glXQp8gcaSxpsj4nOSrgXmI6J9M5gHyFDUnW0rUlq2a3mmbpZV2l4BbY9fMIwxmS2Fi3oL34hhdeVsjw8X9YRvxLC6crbHSy1XvwzCN2JYXTnb48VFPeEbMayunO3xMrbTL+1zjNMnT/LMc4ePO843YljVONvjbSyLeqc5xskTxOSEOHz0xSWevhHDqsbZtrGcfuk0x3j418FLT1zmnWas0pxty3SmntZzWtJLgK8BbwSeBt4TET/Nd6j56TaX+Oyhwzzymd8Z8mjM8uNsW+qZekvP6UuAc4CNks5pO2wT8ExE/DsaDZD+JO+B5smNjKyunG3LcqZ+rOc0gKRmz+nWjQQuA65JPr8T+HNJilH1IGjTfuHowrNP5a6HF47bE9JzjFYlnW4o6rbfqbM9PrLMqaf2nG49JiKOAM8Cr8xjgEvVvHC0cPAQQePC0V0PL/DON854jtEqq1OumzcUXXf5ec72GMtypt6z53QfxyBpM7AZYHZ2NsNbL123Gy/u/9EBHtx60VDGYJa3XjcUPbj1IhfxMZblTD2t5/SiYyQtA14O/LL9hYa1kUAr33hhdeRcWzdZztR79pxO7ADeD3wXeBewc6nz6Xk1IDpteoqFDkH3hSMbBefaipZ6pp7MkX8EuA/4IXBHRDwu6VpJ65PD/gp4paR9wMeBrUsZVLf5wu27FxYds2bbTlZu/QZrtu1c9Fwr7+5iZeFc2zBkWqee1nM6In4F/Ke8BtVrvnDD6pm+us55dxcrC+fahqGUbQLS5gu7/XJcs+PxjiFvfpiN0qC5bnZT7JZts1albBOQdgNFt1+Og4cO9/zT1myUBs11M8vOtmVRyqKeNl+Y9WKQe0ZbmQya6wnJ/dAts1IW9Q2rZ3reQNHpl6MbL/Gyshgk11OTExztspDM2bZOSjmnDvScL+x0kei5F464Z7SVXr+53rJ2Fdfft9fLFy2z0hb1NO2/HO0rB8BLvKx6uhV9Z9uyqmxRb+clXlZXzrb1ozZFHXr/aWtWZc62ZVXKC6VmwyBpnaS9kvZJOu4uaEkfl7RH0qOSvi3p1aMYp1k/XNRtLGXc/GU3MBcRr6OxT8DnhztKs/65qNu4Orb5S0S8ADQ3fzkmIu6PiOeSL3fR6FBqVmqln1PPq6udWZtOm7+8pcfxm4BvdntykL0CnG0rQqmLej8Njsz6lGljFwBJ7wPmgLd1e7GIuAm4CWBubi617bSzbUUp9fRLWoMjsyXIsvkLki4GPgWsj4jn83pzZ9uKUuqi7t1drEDHNn+RdCKNzV92tB4gaTXwlzQK+lN5vrmzbUUpdVFP62pnNqiMm79cD/wG8HVJj0ja0eXl+uZsW1FKXdS9u4sVKSLujYizIuK3IuJzyWOfjogdyecXR8S/iYjzk4/1vV8xO2fbilLqC6W+Pdrqytm2opS6qINvj7b6cratCKWefjEzs/64qJuZ1YiLuplZjbiom5nViIu6mVmNuKibmdWIi7qZWY0oIrWhXDFvLB0Afpbh0OXAvxQ8HI+hOmOAbON4dUScOozBtMuY7Sr9LD2Gco2hZ7ZHVtSzkjQfEXMeg8dQtnEsRVn+HcowDo8h3zF4+sXMrEZc1M3MaqQKRf2mUQ8Aj6GpDGOA8oxjKcry71CGcXgMDbmMofRz6mZmll0VztTNzCwjF3UzsxoZWVGXtE7SXkn7JG3t8PxLJN2ePP89SStanrs6eXyvpLUFjuHjkvZIelTStyW9uuW5o8kWZ0ve5izDOK6UdKDl/T7Q8tz7Jf1j8vH+AsdwQ8v7/1jSwZbncvlZSLpZ0lOSftDleUn6s2SMj0p6Q8tzufwc8uBsZx5D4bnOOI5Csz30XEfE0D+ACeCfgNcAJwLfB85pO+ZDwJeTz68Abk8+Pyc5/iXAyuR1Jgoaw4XAycnn/705huTrfx3iz+JK4M87fO8rgCeSf56SfH5KEWNoO/6jwM0F/Cz+A/AG4Addnr8U+CYg4K3A9/L8OTjb+f33LEOuy5LtYed6VGfqbwb2RcQTEfECcBtwWdsxlwF/nXx+J/B2SUoevy0ino+InwD7ktfLfQwRcX9EPJd8uQs4fYD3WfI4elgL/F1E/DIingH+Dlg3hDFsBG4d4H16iojvAL/scchlwNeiYRcwLelV5PdzyIOznXEMPeT533Pk2R52rkdV1GeAn7d8vT95rOMx0dj5/VnglRm/N68xtNpE4/+mTSdJmpe0S9KGAd6/33G8M/nT7E5JZ/T5vXmNgeTP9JXAzpaH8/pZpOk2zrx+DnlwtvsbQ5G57uu1RpjtXHM9qj1K1eGx9rWV3Y7J8r15jaFxoPQ+YA54W8vDsxHxpKTXADslPRYR/1TQOP4XcGtEPC/pKhpneRdl/N68xtB0BXBnRBxteSyvn0WaojORB2c7+xiKznXWcTSNKtu55mFUZ+r7gTNavj4deLLbMZKWAS+n8SdMlu/NawxIuhj4FLA+Ip5vPh4RTyb/fAJ4AFg9wBgyjSMinm55768Ab+zn3yGPMbS4grY/T3P8WaTpNs68fg55cLYzjmEIue73tUaV7XxzvdSLAANeOFhGY9J/JS9evDi37ZgPs/hi0h3J5+ey+GLSEwx2MSnLGFbTuMhyZtvjpwAvST5fDvwjPS6+5DCOV7V8/vvArnjxQspPkvGcknz+iiLGkBy3CvgpyU1ref8sktdYQfcLSr/L4gtK/5Dnz8HZzu+/ZxlyXaZsDzPXQw99y7/IpcCPk2B9KnnsWhpnDQAnAV+ncbHoH4DXtHzvp5Lv2wtcUuAY/jfw/4BHko8dyeP/HngsCchjwKaCfxbXAY8n73c/cHbL9/635Ge0D/ivRY0h+foaYFvb9+X2s6BxlvQL4DCNs5RNwFXAVcnzAm5MxvgYMJf3z8HZzvW/58hzXYZsDzvXbhNgZlYjvqPUzKxGXNTNzGrERd3MrEZc1M3MasRF3cysRlzUzcxqxEXdzKxG/j8w8DuSGBsrqwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "x = np.linspace(0,1);x # default num=50\n",
    "y = x + np.random.uniform(-0.2, 0.2, x.shape);y\n",
    "# y = x + np.random.normal(-0.2, 0.2, x.shape);y\n",
    "f, ax = plt.subplots(1,2)\n",
    "ax[0].scatter(x, y)\n",
    "ax[1].scatter(x, y+0.3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# get indices of an array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([3.80766021, 2.26049249, 2.97646435, 3.10068452, 4.06445499])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array([ True, False, False,  True,  True])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(array([0, 3, 4]),)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array([3.80766021, 3.10068452, 4.06445499])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = np.random.normal(3, 1, 5);x\n",
    "x > 3\n",
    "np.nonzero(x > 3) # return the index of element where the expression is true\n",
    "x[np.nonzero(x >3)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# bootstrap\n",
    "\n",
    "https://machinelearningmastery.com/a-gentle-introduction-to-the-bootstrap-method/\n",
    "\n",
    "bootstrap is a sampling techincal: random sampling with replacement. so in one bootstrap sample, the one sample may appear 0, 1, or more times. \n",
    "\n",
    "example: draw 3 samples from [0,1,2,3,4,5], bootstrapping works as follow:\n",
    "* drop one sample first, for example, 2, then place 2 back. \n",
    "* draw second sample, for example 5, place 5 back\n",
    "* draw thrid dample, for example 2.\n",
    "\n",
    "so the sample drawn is [2,5,2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 2, 3, 4])"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array([[0, 1, 2, 3, 4],\n",
       "       [5, 6, 7, 8, 9]])"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "[array([0, 1, 2, 3, 4]), array([5, 6, 7, 8, 9])]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "[[0, 1, 2, 3, 4], [5, 6, 7, 8, 9]]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "range(0, 5)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "[0, 1, 2, 3, 4]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.arange(5) # return an array, has no np.range\n",
    "np.arange(10).reshape(2,5)\n",
    "list(np.arange(10).reshape(2,5)) # not recursive\n",
    "np.arange(10).reshape(2,5).tolist() # tolist() is recursive\n",
    "range(5) # range\n",
    "list(range(5)) # a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0  1  2  3\n",
       "0  0  1  2  3\n",
       "1  4  5  6  7"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>[[100, 200, 300], [44, 55, 66]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>[[5, 6, 7]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0  1  2  3                                4\n",
       "0  0  1  2  3  [[100, 200, 300], [44, 55, 66]]\n",
       "1  4  5  6  7                      [[5, 6, 7]]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>[100, 200, 300]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>[44, 55, 66]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>[5, 6, 7]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0  1  2  3                4\n",
       "0  0  1  2  3  [100, 200, 300]\n",
       "0  0  1  2  3     [44, 55, 66]\n",
       "1  4  5  6  7        [5, 6, 7]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>[100, 200, 300]</td>\n",
       "      <td>100</td>\n",
       "      <td>200</td>\n",
       "      <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>[44, 55, 66]</td>\n",
       "      <td>44</td>\n",
       "      <td>55</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>[5, 6, 7]</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0  1  2  3                4    5    6    7\n",
       "0  0  1  2  3  [100, 200, 300]  100  200  300\n",
       "0  0  1  2  3     [44, 55, 66]   44   55   66\n",
       "1  4  5  6  7        [5, 6, 7]    5    6    7"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft = pd.DataFrame(np.arange(8).reshape(2,4))\n",
    "dft.head(2)\n",
    "dft[4] = [[[100, 200, 300], [44,55, 66]], [[5,6, 7]]]\n",
    "dft\n",
    "dft = dft.explode(4)\n",
    "dft\n",
    "dft[[5,6,7]] = pd.DataFrame(dft[4].values.tolist(), index=dft.index)\n",
    "dft"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>200</td>\n",
       "      <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>44</td>\n",
       "      <td>55</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     0    1    2\n",
       "0  100  200  300\n",
       "0   44   55   66\n",
       "1    5    6    7"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(dft[4].values.tolist(), index=dft.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### read_csv, columns, shape, and head\n",
    "read demo file into a pandas dataframe, show column headers, number of rows and columns of the dataframe, show first 2 rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chr', 'start', 'end', 'patients', 'num_strelka_patients', 'gene',\n",
       "       'feature_of_interest', 'num_normal_patients', 'num_mutationseq_patient',\n",
       "       'Reference_Allele', 'Tumor_Seq_Allele2', 'cum_pos', 'feature_colors',\n",
       "       'Clade', 'A7_count', 'A9_count'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(174, 16)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>patients</th>\n",
       "      <th>num_strelka_patients</th>\n",
       "      <th>gene</th>\n",
       "      <th>feature_of_interest</th>\n",
       "      <th>num_normal_patients</th>\n",
       "      <th>num_mutationseq_patient</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>cum_pos</th>\n",
       "      <th>feature_colors</th>\n",
       "      <th>Clade</th>\n",
       "      <th>A7_count</th>\n",
       "      <th>A9_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>183675458</td>\n",
       "      <td>183675459</td>\n",
       "      <td>S_99020;S_99120;S_99195</td>\n",
       "      <td>3</td>\n",
       "      <td>ABCC5</td>\n",
       "      <td>active_enhancer</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>676125452</td>\n",
       "      <td>green</td>\n",
       "      <td>A7;A9;A7</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>19268412</td>\n",
       "      <td>19268413</td>\n",
       "      <td>S_99120;S_99170;S_99197;S_99326;S_99354;S_9939...</td>\n",
       "      <td>8</td>\n",
       "      <td>ABHD3</td>\n",
       "      <td>active_enhancer</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>2600635486</td>\n",
       "      <td>green</td>\n",
       "      <td>A9;A7;A9;A7;A7;A9;A9;A9</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr      start        end  \\\n",
       "0    3  183675458  183675459   \n",
       "1   18   19268412   19268413   \n",
       "\n",
       "                                            patients  num_strelka_patients  \\\n",
       "0                            S_99020;S_99120;S_99195                     3   \n",
       "1  S_99120;S_99170;S_99197;S_99326;S_99354;S_9939...                     8   \n",
       "\n",
       "    gene feature_of_interest  num_normal_patients  num_mutationseq_patient  \\\n",
       "0  ABCC5     active_enhancer                    0                        5   \n",
       "1  ABHD3     active_enhancer                    0                        8   \n",
       "\n",
       "  Reference_Allele Tumor_Seq_Allele2     cum_pos feature_colors  \\\n",
       "0                C                 T   676125452          green   \n",
       "1                T                 A  2600635486          green   \n",
       "\n",
       "                     Clade  A7_count  A9_count  \n",
       "0                 A7;A9;A7         2         1  \n",
       "1  A9;A7;A9;A7;A7;A9;A9;A9         3         5  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = '/projects/trans_scratch/validations/workspace/szong/Cervical/notebook/demo.csv'\n",
    "df = pd.read_csv(f)\n",
    "df.columns\n",
    "df.shape\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### np.where\n",
    "annotate rows based on the values in relationship between two columns. For example, if A7_count > A9_count, annotate the entry as red; otherwise assign a color of blue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>patients</th>\n",
       "      <th>num_strelka_patients</th>\n",
       "      <th>gene</th>\n",
       "      <th>feature_of_interest</th>\n",
       "      <th>num_normal_patients</th>\n",
       "      <th>num_mutationseq_patient</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>cum_pos</th>\n",
       "      <th>feature_colors</th>\n",
       "      <th>Clade</th>\n",
       "      <th>A7_count</th>\n",
       "      <th>A9_count</th>\n",
       "      <th>clade_colors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>183675458</td>\n",
       "      <td>183675459</td>\n",
       "      <td>S_99020;S_99120;S_99195</td>\n",
       "      <td>3</td>\n",
       "      <td>ABCC5</td>\n",
       "      <td>active_enhancer</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>676125452</td>\n",
       "      <td>green</td>\n",
       "      <td>A7;A9;A7</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>red</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18</td>\n",
       "      <td>19268412</td>\n",
       "      <td>19268413</td>\n",
       "      <td>S_99120;S_99170;S_99197;S_99326;S_99354;S_9939...</td>\n",
       "      <td>8</td>\n",
       "      <td>ABHD3</td>\n",
       "      <td>active_enhancer</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>2600635486</td>\n",
       "      <td>green</td>\n",
       "      <td>A9;A7;A9;A7;A7;A9;A9;A9</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>blue</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr      start        end  \\\n",
       "0    3  183675458  183675459   \n",
       "1   18   19268412   19268413   \n",
       "\n",
       "                                            patients  num_strelka_patients  \\\n",
       "0                            S_99020;S_99120;S_99195                     3   \n",
       "1  S_99120;S_99170;S_99197;S_99326;S_99354;S_9939...                     8   \n",
       "\n",
       "    gene feature_of_interest  num_normal_patients  num_mutationseq_patient  \\\n",
       "0  ABCC5     active_enhancer                    0                        5   \n",
       "1  ABHD3     active_enhancer                    0                        8   \n",
       "\n",
       "  Reference_Allele Tumor_Seq_Allele2     cum_pos feature_colors  \\\n",
       "0                C                 T   676125452          green   \n",
       "1                T                 A  2600635486          green   \n",
       "\n",
       "                     Clade  A7_count  A9_count clade_colors  \n",
       "0                 A7;A9;A7         2         1          red  \n",
       "1  A9;A7;A9;A7;A7;A9;A9;A9         3         5         blue  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['clade_colors'] = np.where((df.A7_count>df.A9_count), 'red', 'blue')\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pd.crosstab\n",
    "crosstab can calculate frequency\n",
    "\n",
    "1) we can count the frequency of all chromosome and hotspot type combinations\n",
    "\n",
    "2) we can also calculate the relative frequencies across rows, columns, or all."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.crosstab??"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>feature_of_interest</th>\n",
       "      <th>active_enhancer</th>\n",
       "      <th>other</th>\n",
       "      <th>potential_promoter</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "feature_of_interest  active_enhancer  other  potential_promoter\n",
       "chr                                                            \n",
       "1                                  2     12                   0\n",
       "2                                  0     13                   1"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>feature_of_interest</th>\n",
       "      <th>active_enhancer</th>\n",
       "      <th>other</th>\n",
       "      <th>potential_promoter</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.094203</td>\n",
       "      <td>0.083333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "feature_of_interest  active_enhancer     other  potential_promoter\n",
       "chr                                                               \n",
       "1                           0.083333  0.086957            0.000000\n",
       "2                           0.000000  0.094203            0.083333"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>feature_of_interest</th>\n",
       "      <th>active_enhancer</th>\n",
       "      <th>other</th>\n",
       "      <th>potential_promoter</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.011494</td>\n",
       "      <td>0.068966</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.074713</td>\n",
       "      <td>0.005747</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "feature_of_interest  active_enhancer     other  potential_promoter\n",
       "chr                                                               \n",
       "1                           0.011494  0.068966            0.000000\n",
       "2                           0.000000  0.074713            0.005747"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>feature_of_interest</th>\n",
       "      <th>active_enhancer</th>\n",
       "      <th>other</th>\n",
       "      <th>potential_promoter</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "feature_of_interest  active_enhancer  other  potential_promoter\n",
       "chr                                                            \n",
       "1                                2.0   12.0                 NaN\n",
       "2                                NaN   13.0                 1.0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.crosstab(index=df['chr'], columns=df['feature_of_interest']).head(2)\n",
    "pd.crosstab(index=df['chr'], columns=df['feature_of_interest'], normalize='columns').head(2)\n",
    "pd.crosstab(index=df['chr'], columns=df['feature_of_interest'], normalize='all').head(2)\n",
    "df.pivot_table(index='chr', columns='feature_of_interest', aggfunc='size').head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### merge two dataframe based on multiple columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df3' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-71-1bb29f92fca5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m dff = pd.merge(df, df3, left_on=['gene', 'chr', 'start_x', 'Reference_Allele', 'Tumor_Seq_Allele2'], \n\u001b[0m\u001b[1;32m      2\u001b[0m                right_on=['gene', 'chr', 'start_x', 'Reference_Allele', 'Tumor_Seq_Allele2'], how='left')\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df3' is not defined"
     ]
    }
   ],
   "source": [
    "dff = pd.merge(df, df3, left_on=['gene', 'chr', 'start_x', 'Reference_Allele', 'Tumor_Seq_Allele2'], \n",
    "               right_on=['gene', 'chr', 'start_x', 'Reference_Allele', 'Tumor_Seq_Allele2'], how='left')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### modify values in a column, use map or regex\n",
    " map a column to a new column with different values based on a predifined dictionary \n",
    " \n",
    " regex can be very powerful: https://docs.python.org/2/library/re.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['indel_colors'] = df.KRAS_snvs_indels.map({'N':'grey', 'Y':'red', 'germline':'black'})\n",
    "dfn['patients_new'] = list(map(lambda x: x.replace('_T', ''), s))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cov.loc['clinical_stage',:] = df_cov.loc['clinical_stage',:].str.replace('[AB]\\d*', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# wide to long transformation, break entries in a column into multiple rows \n",
    "# string split to expand a dataframe to long form by splitting values in a column\n",
    "s = df.patients.str.split(';', expand=True).stack()\n",
    "idx = s.index.get_level_values(0)\n",
    "dfn = df.loc[idx]\n",
    "dfn['new_col'] = list(s)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['cumsum_num_strelka'] = df.sort_values(['chr', 'start_x']).num_strela_patients.cumsum()\n",
    "# cumulative sum is useful, for example calculate the cumulative genome positions when plotting events along the genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = np.argmax(p_matrix, axis=1)\n",
    "# get the index of max value in a ndarray"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### groupby and aggregate\n",
    "groupby_aggregate is a powerful way to manipulate data. For instance, to get the mean, minimum, maxmum, and standard deviation of the aggregated series. The syntax is df.groupby(['group col1', 'group col2].agg({'aggregate col1':[min, max, std], 'aggregate col2':[agg_func1, agg_func2]}). This gives you a both row and column multiindexed dataframe. you could rename the columns and reset the index to flatten the index so that following manipulation is easier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df3 = df2.astype(str).groupby(['gene', 'chr', 'start_x', 'Reference_Allele', 'Tumor_Seq_Allele2']).\\\n",
    "agg({'hiv_status':[';'.join, count_pos, count_neg], 'previous_final_hist':';'.join, \\\n",
    "     'Clade':[';'.join, count_A7, count_A9]}).reset_index()\n",
    "df3.head(3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2['barcode'] = df2[['Sentrix Barcode', 'Sample Section']].astype(str).apply(lambda x: '_'.join(x), axis=1) \n",
    "dfy['region'] = dfy[['Chrom', 'left', 'right']].astype(str).apply(lambda x: x[0]+':'+x[1]+'-'+x[2], axis=1)\n",
    "# join two columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_dict = pd.Series(df2.sample_id.values,index=df2.barcode).to_dict()\n",
    "# fastest way to convert two df columns to a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfy = dfy[~dfy.pos.astype(str).str.contains('\\(')]\n",
    "# drop rows based on values in a column that matches a pattern\n",
    "# replace only a few values in a column, you can give a list of things you would like to be replaced\n",
    "df['chr'] = df['chr'].replace(['X'], 23)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "pivot, pivot_table, and crosstab, crosstab is a function, not a pandas method\n",
    "1) pivot only reshapes dataframe, does not aggregate. can always be replaced by pivot_table\n",
    "\n",
    "2) always use pivot_table, need to provide aggfunc even if you just want to reshape the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alphas</th>\n",
       "      <th>species</th>\n",
       "      <th>prevalence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>[0.1 0.1 0.1]</td>\n",
       "      <td>lions</td>\n",
       "      <td>0.492063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>[1 1 1]</td>\n",
       "      <td>lions</td>\n",
       "      <td>0.444444</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          alphas species  prevalence\n",
       "0  [0.1 0.1 0.1]   lions    0.492063\n",
       "1        [1 1 1]   lions    0.444444"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alphas</th>\n",
       "      <th>species</th>\n",
       "      <th>prevalence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>[0.1 0.1 0.1]</td>\n",
       "      <td>lions</td>\n",
       "      <td>0.492063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>[1 1 1]</td>\n",
       "      <td>lions</td>\n",
       "      <td>0.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>[5 5 5]</td>\n",
       "      <td>lions</td>\n",
       "      <td>0.380952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>[15 15 15]</td>\n",
       "      <td>lions</td>\n",
       "      <td>0.352941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>[0.1 0.1 0.1]</td>\n",
       "      <td>tigers</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>[1 1 1]</td>\n",
       "      <td>tigers</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>[5 5 5]</td>\n",
       "      <td>tigers</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>[15 15 15]</td>\n",
       "      <td>tigers</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>[0.1 0.1 0.1]</td>\n",
       "      <td>bears</td>\n",
       "      <td>0.174603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>[1 1 1]</td>\n",
       "      <td>bears</td>\n",
       "      <td>0.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>[5 5 5]</td>\n",
       "      <td>bears</td>\n",
       "      <td>0.285714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>[15 15 15]</td>\n",
       "      <td>bears</td>\n",
       "      <td>0.313725</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           alphas species  prevalence\n",
       "0   [0.1 0.1 0.1]   lions    0.492063\n",
       "1         [1 1 1]   lions    0.444444\n",
       "2         [5 5 5]   lions    0.380952\n",
       "3      [15 15 15]   lions    0.352941\n",
       "4   [0.1 0.1 0.1]  tigers    0.333333\n",
       "5         [1 1 1]  tigers    0.333333\n",
       "6         [5 5 5]  tigers    0.333333\n",
       "7      [15 15 15]  tigers    0.333333\n",
       "8   [0.1 0.1 0.1]   bears    0.174603\n",
       "9         [1 1 1]   bears    0.222222\n",
       "10        [5 5 5]   bears    0.285714\n",
       "11     [15 15 15]   bears    0.313725"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def times100(x):\n",
    "    return 100*x\n",
    "df = pd.read_csv('/projects/trans_scratch/validations/workspace/szong/Cervical/notebook/animals_long.txt')\n",
    "df.head(2)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>species</th>\n",
       "      <th>alphas</th>\n",
       "      <th>bears</th>\n",
       "      <th>lions</th>\n",
       "      <th>tigers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>[0.1 0.1 0.1]</td>\n",
       "      <td>0.174603</td>\n",
       "      <td>0.492063</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>[1 1 1]</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>[15 15 15]</td>\n",
       "      <td>0.313725</td>\n",
       "      <td>0.352941</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>[5 5 5]</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "species         alphas     bears     lions    tigers\n",
       "0        [0.1 0.1 0.1]  0.174603  0.492063  0.333333\n",
       "1              [1 1 1]  0.222222  0.444444  0.333333\n",
       "2           [15 15 15]  0.313725  0.352941  0.333333\n",
       "3              [5 5 5]  0.285714  0.380952  0.333333"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = df.pivot(index='alphas', columns='species', values='prevalence').reset_index()\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alphas</th>\n",
       "      <th>animals</th>\n",
       "      <th>ori_prevalence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>[0.1 0.1 0.1]</td>\n",
       "      <td>bears</td>\n",
       "      <td>0.174603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>[1 1 1]</td>\n",
       "      <td>bears</td>\n",
       "      <td>0.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>[15 15 15]</td>\n",
       "      <td>bears</td>\n",
       "      <td>0.313725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>[5 5 5]</td>\n",
       "      <td>bears</td>\n",
       "      <td>0.285714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>[0.1 0.1 0.1]</td>\n",
       "      <td>lions</td>\n",
       "      <td>0.492063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>[1 1 1]</td>\n",
       "      <td>lions</td>\n",
       "      <td>0.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>[15 15 15]</td>\n",
       "      <td>lions</td>\n",
       "      <td>0.352941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>[5 5 5]</td>\n",
       "      <td>lions</td>\n",
       "      <td>0.380952</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          alphas animals  ori_prevalence\n",
       "0  [0.1 0.1 0.1]   bears        0.174603\n",
       "1        [1 1 1]   bears        0.222222\n",
       "2     [15 15 15]   bears        0.313725\n",
       "3        [5 5 5]   bears        0.285714\n",
       "4  [0.1 0.1 0.1]   lions        0.492063\n",
       "5        [1 1 1]   lions        0.444444\n",
       "6     [15 15 15]   lions        0.352941\n",
       "7        [5 5 5]   lions        0.380952"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.melt(df2, id_vars='alphas', value_vars=['bears', 'lions'], value_name='ori_prevalence', var_name='animals')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>species</th>\n",
       "      <th>bears</th>\n",
       "      <th>lions</th>\n",
       "      <th>tigers</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alphas</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>[0.1 0.1 0.1]</td>\n",
       "      <td>0.174603</td>\n",
       "      <td>0.492063</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>[1 1 1]</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>[15 15 15]</td>\n",
       "      <td>0.313725</td>\n",
       "      <td>0.352941</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>[5 5 5]</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "species           bears     lions    tigers\n",
       "alphas                                     \n",
       "[0.1 0.1 0.1]  0.174603  0.492063  0.333333\n",
       "[1 1 1]        0.222222  0.444444  0.333333\n",
       "[15 15 15]     0.313725  0.352941  0.333333\n",
       "[5 5 5]        0.285714  0.380952  0.333333"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pivot a table, pivot_table?\n",
    "df3 = df.pivot_table(index='alphas', columns='species', values='prevalence', aggfunc=lambda x:x)# aggfunc=times100)\n",
    "df3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient</th>\n",
       "      <th>vcf</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>HTMCP-03-06-02025</td>\n",
       "      <td>/projects/somatic/NCI_SAIC_HIV_Cervical/HTMCP-...</td>\n",
       "      <td>indels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>HTMCP-03-06-02025</td>\n",
       "      <td>/projects/somatic/NCI_SAIC_HIV_Cervical/HTMCP-...</td>\n",
       "      <td>snvs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             patient                                                vcf  \\\n",
       "0  HTMCP-03-06-02025  /projects/somatic/NCI_SAIC_HIV_Cervical/HTMCP-...   \n",
       "1  HTMCP-03-06-02025  /projects/somatic/NCI_SAIC_HIV_Cervical/HTMCP-...   \n",
       "\n",
       "     type  \n",
       "0  indels  \n",
       "1    snvs  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### pivot_table for non-numerical values, long to wide transformation\n",
    "df1 = pd.read_csv('/projects/trans_scratch/validations/workspace/szong/Cervical/variant_bwamem/validation/strelka_classic_vcf_files.txt.test')\n",
    "df1.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient</th>\n",
       "      <th>indels</th>\n",
       "      <th>snvs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>HTMCP-03-06-02025</td>\n",
       "      <td>/projects/somatic/NCI_SAIC_HIV_Cervical/HTMCP-...</td>\n",
       "      <td>/projects/somatic/NCI_SAIC_HIV_Cervical/HTMCP-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>HTMCP-03-06-02044</td>\n",
       "      <td>/projects/somatic/NCI_SAIC_HIV_Cervical/HTMCP-...</td>\n",
       "      <td>/projects/somatic/NCI_SAIC_HIV_Cervical/HTMCP-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             patient                                             indels  \\\n",
       "0  HTMCP-03-06-02025  /projects/somatic/NCI_SAIC_HIV_Cervical/HTMCP-...   \n",
       "1  HTMCP-03-06-02044  /projects/somatic/NCI_SAIC_HIV_Cervical/HTMCP-...   \n",
       "\n",
       "                                                snvs  \n",
       "0  /projects/somatic/NCI_SAIC_HIV_Cervical/HTMCP-...  \n",
       "1  /projects/somatic/NCI_SAIC_HIV_Cervical/HTMCP-...  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# both works, pivot_table for non_numerical values, long to wide transformation\n",
    "# df2 = pd.pivot_table(df1, index='patient', columns='type', values='vcf', aggfunc=lambda x: str(x.values[0])).reset_index()\n",
    "df2 = pd.pivot_table(df1, index='patient', columns='type', values='vcf', aggfunc=lambda x: ''.join([str(v) for v in x])).reset_index()\n",
    "df2.columns = ['patient', 'indels', 'snvs']\n",
    "df2.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chr', 'start', 'end', 'patients', 'num_strelka_patients', 'gene',\n",
       "       'feature_of_interest', 'num_normal_patients', 'num_mutationseq_patient',\n",
       "       'Reference_Allele', 'Tumor_Seq_Allele2', 'cum_pos', 'feature_colors',\n",
       "       'Clade', 'A7_count', 'A9_count'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(174, 16)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>patients</th>\n",
       "      <th>num_strelka_patients</th>\n",
       "      <th>gene</th>\n",
       "      <th>feature_of_interest</th>\n",
       "      <th>num_normal_patients</th>\n",
       "      <th>num_mutationseq_patient</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>cum_pos</th>\n",
       "      <th>feature_colors</th>\n",
       "      <th>Clade</th>\n",
       "      <th>A7_count</th>\n",
       "      <th>A9_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>183675458</td>\n",
       "      <td>183675459</td>\n",
       "      <td>S_99020;S_99120;S_99195</td>\n",
       "      <td>3</td>\n",
       "      <td>ABCC5</td>\n",
       "      <td>active_enhancer</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>676125452</td>\n",
       "      <td>green</td>\n",
       "      <td>A7;A9;A7</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18</td>\n",
       "      <td>19268412</td>\n",
       "      <td>19268413</td>\n",
       "      <td>S_99120;S_99170;S_99197;S_99326;S_99354;S_9939...</td>\n",
       "      <td>8</td>\n",
       "      <td>ABHD3</td>\n",
       "      <td>active_enhancer</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>2600635486</td>\n",
       "      <td>green</td>\n",
       "      <td>A9;A7;A9;A7;A7;A9;A9;A9</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr      start        end  \\\n",
       "0    3  183675458  183675459   \n",
       "1   18   19268412   19268413   \n",
       "\n",
       "                                            patients  num_strelka_patients  \\\n",
       "0                            S_99020;S_99120;S_99195                     3   \n",
       "1  S_99120;S_99170;S_99197;S_99326;S_99354;S_9939...                     8   \n",
       "\n",
       "    gene feature_of_interest  num_normal_patients  num_mutationseq_patient  \\\n",
       "0  ABCC5     active_enhancer                    0                        5   \n",
       "1  ABHD3     active_enhancer                    0                        8   \n",
       "\n",
       "  Reference_Allele Tumor_Seq_Allele2     cum_pos feature_colors  \\\n",
       "0                C                 T   676125452          green   \n",
       "1                T                 A  2600635486          green   \n",
       "\n",
       "                     Clade  A7_count  A9_count  \n",
       "0                 A7;A9;A7         2         1  \n",
       "1  A9;A7;A9;A7;A7;A9;A9;A9         3         5  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df2 = pd.pivot_table(df1, index='patient', columns='type', values='vcf', aggfunc=lambda x: ''.join([str(v) for v in x])).reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### melt and stack\n",
    "1) melt and stack reshape data in exactly the same way.\n",
    "\n",
    "2) melt: two parameters, id_vars which are the column names that are to remain vertical (and not reshaped) and value_vars which are the column names to be reshaped into a single column.\n",
    "\n",
    "3) use melt over stack because melt allows you to rename columns and avoids multiindex.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      lions    tigers     bears         alphas\n",
      "0  0.492063  0.333333  0.174603  [0.1 0.1 0.1]\n",
      "1  0.444444  0.333333  0.222222        [1 1 1]\n",
      "2  0.380952  0.333333  0.285714        [5 5 5]\n",
      "3  0.352941  0.333333  0.313725     [15 15 15]\n",
      "           alphas species  prevalence\n",
      "0   [0.1 0.1 0.1]   lions    0.492063\n",
      "1         [1 1 1]   lions    0.444444\n",
      "2         [5 5 5]   lions    0.380952\n",
      "3      [15 15 15]   lions    0.352941\n",
      "4   [0.1 0.1 0.1]  tigers    0.333333\n",
      "5         [1 1 1]  tigers    0.333333\n",
      "6         [5 5 5]  tigers    0.333333\n",
      "7      [15 15 15]  tigers    0.333333\n",
      "8   [0.1 0.1 0.1]   bears    0.174603\n",
      "9         [1 1 1]   bears    0.222222\n",
      "10        [5 5 5]   bears    0.285714\n",
      "11     [15 15 15]   bears    0.313725\n"
     ]
    }
   ],
   "source": [
    "# melt wide form to long form, it is like unpivot. pivot long to wide\n",
    "value_df = pd.read_csv('/projects/trans_scratch/validations/workspace/szong/Cervical/notebook/animals.txt')\n",
    "print(value_df)\n",
    "melted = pd.melt(value_df, id_vars = 'alphas', value_vars=['lions', 'tigers', 'bears'], value_name='prevalence',\n",
    "        var_name = 'species')\n",
    "print(melted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "alphas               \n",
       "[0.1 0.1 0.1]  lions     0.492063\n",
       "               tigers    0.333333\n",
       "               bears     0.174603\n",
       "[1 1 1]        lions     0.444444\n",
       "               tigers    0.333333\n",
       "               bears     0.222222\n",
       "[5 5 5]        lions     0.380952\n",
       "               tigers    0.333333\n",
       "               bears     0.285714\n",
       "[15 15 15]     lions     0.352941\n",
       "               tigers    0.333333\n",
       "               bears     0.313725\n",
       "dtype: float64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# use stack can achieve identical results, stack works with index\n",
    "value_df.set_index('alphas').stack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfsAAAGHCAYAAAC3a/toAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAIABJREFUeJzt3XmUnGWd//33lxBICDuEYRMiIEs2Gg2bLAYZQMWFAR3wASWgsgSCgUEBZZT5wYPLoxgYgxk4DiGscdDx8FNQhBGRTUhMA02CDEuIhC1BAlnJ9n3+uKtj0+mlklR1p+9+v86pk657ua6r6q7Up+7tuiIzkSRJ5bVBdzdAkiTVl2EvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kuomIkZFxIM9rWypbAx7aT0SETMjYnFELGjx+HE3tuf+iPhyHcrtFxHzIuKjbcz7UUTcUes6pd5sw+5ugKTVfCoz7+3uRtRTZi6JiMnAF4H/aZ4eEX2AzwNf6a62SWXknr3UQ0TET1ru8UbE9yLiviiMjIiXI+IbETG3coTg5BbLbhwRP4iIWRHxekRMiIj+LeZ/JiIaI+KdiHg+Ij4WEf8vcBjw45ZHGCJi74j4XUT8LSL+EhH/3KKcbSLizko5jwG7d/CSbgROiIhNWkw7huJ76e5KeRdX2jM/IqZHxD+1894MioiMiA1bTHvPUYmIOD0iZkTEWxHx24jYtbP3XCoLw17qOf4FGF45V30Y8CXg1Px7n9fbA9sCOwGnAtdFxF6Ved8D9gQagD0qy3wLICIOACYBXwO2BA4HZmbmN4E/Audm5qaZeW5EDAB+B9wKbEexF35tRAyp1DMeWALsAJxeebQpMx8GXgWObzH5C8Ctmbm88vx5ih8cWwD/BtwcETtU/5YVIuI44BuVugZWXtdta1qO1FMZ9tL655eV89nNj68AZOYi4BTgKuBmYExmvtxq3X/NzHcz8w/Ar4F/joigOCx+fmb+LTPnA1cCJ1XW+RLwn5n5u8xcmZmzM/OZdtr2SYofAjdk5vLM/DPwc+CzlUPwJwDfysyFmdlEsffekUkUh/KJiM2Bz7RcJzP/KzNfqbRrMvC/wAGdlNmWM4HvZOaMyg+JK4EG9+7VW3jOXlr/HNfeOfvMfCwiXqDYq/5Zq9lvZebCFs9fAnak2JPdBJha5D4AAfSp/P0+4K4q27YrcGBEzGsxbUPgpko9GwJ/bdWGjkwCvh0RO1Ecwn8uM6etamTEF4ELgEGVSZtSHL1YU7sCV0fED1tMC4ojHJ21Uerx3LOXepCIOAfYGHgF+Hqr2VtVDrM326Wy3FxgMTAkM7esPLbIzE0ry/2V9s+ttx4W86/AH1qUs2XlEP/ZwBxgOcWPh5ZtaFdmzqI4pH4yxSH8SS1e667A9cC5wDaZuSXQRBHSrTX/yGl5/n/7Vu0+s1W7+1dOJUilZ9hLPURE7AlcQXEo/wvA1yOiodVi/xYRG1XO6X8S+K/MXEkRmj+KiO0qZe0UEcdU1vkpcFpEHBkRG1Tm7V2Z9zqwW4vyfwXsGRFfiIi+lcf+EbFPZq4AfgFcFhGbRMRgimsHOnMjRaAfAtzSYvoAih8bcyptPg0Y2lYBmTkHmA2cEhF9IuJ03vsDZgJwSfO1BRGxRUR8roq2SaVg2Evrn//b6j77/65cZX4z8L3MfCIz/5figrObImLjynqvAW9R7M3fApzV4tz7RcBzwKMR8Q5wL7AXFKcGgNOAHwFvA3+gOOwNcDXF+fi3IuKayvn+oynO979SqfN7FEcboAjtTSvTJwI3VPF67wC2Au7LzFebJ2bmdOCHwCMUPzqGAQ91UM5XKC4yfBMYAqzaa8/M/6608/bK628CPl5F26RSiL9fyCupp4qIkcDNmblzd7dF0vrHPXtJkkrOsJckqeQ8jC9JUsm5Zy9JUskZ9pIklVxpetDbdtttc9CgQd3dDEmSuszUqVPnZubAzpYrTdgPGjSIKVOmdHczJEnqMhFRVXfPHsaXJKnkDHtJkkrOsJckqeRKc85eklRey5Yt4+WXX2bJkiXd3ZRu0a9fP3beeWf69u27VuvXPewjYjTF4BQ7AE8DYzPzj+0sOxL4fRuz9mkxoIckqZd5+eWX2WyzzRg0aBARbY1yXF6ZyZtvvsnLL7/M+9///rUqo66H8SPiRIpRs64E9qMYheruiOhwjGuKEat2aPH433q2U5K0fluyZAnbbLNNrwt6gIhgm222WaejGvU+Z38BMDEzr8/MGZk5BngVOLuT9d7IzNdaPFbUuZ2SpPVcbwz6Zuv62usW9hGxEfAh4J5Ws+4BPtzJ6lMi4tWIuC8ijqhLAyVJWgef+MQnmDdvXnc3oyr13LPfFugDvN5q+uvA9u2s07zXfwJwPPAX4L6IOLythSPijIiYEhFT5syZU5tWS5JUhbvuuostt9yyu5tRla649a71sHrRxrRiwcy/ZOaEzJyamY9k5mjgN8CF7Sx/XWaOyMwRAwd22lugJKmXWbhwIcceeyz77rsvQ4cOZfLkyQwaNIiLLrqIAw44gAMOOIDnnnsOgDlz5nDCCSew//77s//++/PQQw8BsGDBAk477TSGDRvG8OHD+fnPfw4UPbfOnTsXgJtvvpkDDjiAhoYGzjzzTFasWMGKFSsYNWoUQ4cOZdiwYfzoRz/qnjeB+l6NPxdYwep78dux+t5+R/4EnFSrRkmSeo/f/OY37Ljjjvz6178G4O233+aiiy5i880357HHHmPSpEmMHTuWX/3qV3z1q1/l/PPP59BDD2XWrFkcc8wxzJgxg8svv5wtttiCp556CoC33nrrPXXMmDGDyZMn89BDD9G3b19Gjx7NLbfcwpAhQ5g9ezZNTU0A3XrIv25hn5lLI2IqcBTwXy1mHQX8fA2KaqA4vC9J0hoZNmwYF154IRdddBGf/OQnOeywwwD4/Oc/v+rf888/H4B7772X6dOnr1r3nXfeYf78+dx7773cfvvtq6ZvtdVW76njvvvuY+rUqey///4ALF68mO22245PfepTvPDCC4wZM4Zjjz2Wo48+uq6vtSP1vs/+KuCmiHgMeAg4C9gRmAAQEZMAMvOLledjgZkU9+NvBJwCHEdxDr8mxo4dS2NjY62KWyMNDQ2MGzeuW+qWpN5ozz33ZOrUqdx1111ccsklqwK35dXtzX+vXLmSRx55hP79+7+njMzs8Gr4zOTUU0/lO9/5zmrznnjiCX77298yfvx4fvazn/Gf//mftXhZa6yuYZ+ZkyNiG+BSivvlm4BPZGbzKD2t77ffCPgBsBOwmCL0j83Mu2rVpsbGRh589HH6D+zsVv/aWjxnVpfWJ0mCV155ha233ppTTjmFTTfdlIkTJwIwefJkLr74YiZPnszBBx8MwNFHH82Pf/xjvva1rwFFXjQ0NKya3ryz9tZbb71n7/7II4/kM5/5DOeffz7bbbcdf/vb35g/fz4DBgxgo4024oQTTmD33Xdn1KhRXfraW6p7D3qZeS1wbTvzRrZ6/n3g+/VuU/+Bu7DnSZfUu5r3ePb21X/xSZLq66mnnuJrX/saG2ywAX379uUnP/kJn/3sZ3n33Xc58MADWblyJbfddhsA11xzDeeccw7Dhw9n+fLlHH744UyYMIFLL72Uc845h6FDh9KnTx++/e1vc/zxx6+qY/DgwVxxxRUcffTRrFy5kr59+zJ+/Hj69+/PaaedxsqVKwHa3PPvKpHZ5oXxPc6IESOymvHsR44cydTnX++WsP/Q7v/A/fff36X1SlIZzJgxg3322acmZQ0aNIgpU6aw7bbb1qS8rtLWexARUzNzRGfrOuqdJEkl56h3kqReZebMmd3dhC7nnr0kSSVn2EuSVHKGvSRJJWfYS5JUcl6gJ0nqUerRE2o1PZxuuummLFiwgFdeeYXzzjuPO+64o6ZtqCfDXpLUo9S6J9Q17eF0xx137FFBD4a9JKkHqmVPqGvaw+nMmTP55Cc/SVNTE0uWLOHss89mypQpbLjhhlx11VUcccQRTJw4kTvvvJNFixbx/PPP80//9E98//vfZ8WKFXzpS19iypQpRASnn376qoF46smwlyRpLY0fPx4ouuV95plnOProo3n22WeB4gjEtGnT2Hjjjdlrr70YM2YMb7zxRrcMe+sFepIkraUHH3yQL3zhCwDsvffe7LrrrqvC/sgjj2SLLbagX79+DB48mJdeeonddttt1bC3v/nNb9h88827pJ2GvSRJa6mj8WU23njjVX/36dOH5cuXs9VWW/HEE08wcuRIxo8fz5e//OWuaKZhL0nS2jr88MO55ZZbAHj22WeZNWsWe+21V7vLz507l5UrV3LCCSdw+eWX8+c//7lL2uk5e0lSj7N4zqyaDR2+eM4s2P0f1mrd0aNHc9ZZZzFs2DA23HBDJk6c+J49+tZmz57dLcPeGvaSpB6loaGhtgXu/g9VlblgwQKgGCK3+QK7fv36MXHixNWWHTVqFKNGjVr1/Fe/+tWqv7tqb74lw16S1KN01vmNVuc5e0mSSs6wlySp5Ax7SZJKzrCXJKnkDHtJkkrOq/ElST1KdwxxO2/ePG699VZGjx7tELeSJNVbY2Mj0/70RwZv368m5U1/bUmny8ybN49rr72W0aNH13SI2xUrVtCnT5+alNURw16S1OMM3r4fk0/brSZlnXjDC50uc/HFF/P888/T0NDABz7wAWbMmEFTUxOLFi1i1KhRPPPMM+yzzz7MnDmT8ePHM2LECO655x6+/e1v8+6777L77rtzww03sOmmmzJo0CBOP/107rnnHs4991zeeOMNJkyYwIYbbsjgwYO5/fbba/K6WjLsJUnqxHe/+12amppobGxcNZ49wLXXXstWW23Fk08+SVNT06qe+ObOncsVV1zBvffey4ABA/je977HVVddxbe+9S2g6HnvwQcfBGDHHXfkxRdfZOONN67bkLeGvSRJa+nBBx/kq1/9KgBDhw5l+PDhADz66KNMnz6dQw45BIClS5dy8MEHr1rvxBNPXPX38OHDOfnkkznuuOM47rjj6tJOw149Sj0uzKlWZxfwSOp92hviNjM56qijuO2229qcP2DAgFV///rXv+aBBx7gzjvv5PLLL+fpp59mww1rG8+GvXqUxsZGHnz0cfoP3KVL6108Z1aX1idp/bLZZpsxf/781aYfeuih/OxnP+OII45g+vTpPPXUUwAcdNBBnHPOOTz33HPsscceLFq0iJdffpk999zzPeuvXLmSv/71rxxxxBEceuih3HrrrSxYsIAtt9yypu037NXj9B+4C3uedEmX1lmroTQl1cb015ZUdWFdtWXtt2vHy2yzzTYccsghDB06lH322WfV9NGjR3PqqacyfPhw9ttvP4YPH84WW2zBwIEDmThxIp///Od59913AbjiiitWC/sVK1Zwyimn8Pbbb5OZnH/++TUPejDsJUk9TK2HuN1v1+rKvPXWW1eb1q9fP26++Wb69evH888/z5FHHsmuuxa/HD760Y/y+OOPr7bOzJkzV/3dt2/fVRfq1ZNhL0nqUdana2cWLVrEEUccwbJly8hMfvKTn7DRRht1d7NWY9hLkrSWNttsM6ZMmdLdzeiUfeNLklRyhr0kSSVn2EuSVHKGvSRJJecFepKkHqU7hrgFVvWJ39TUVNO6u4JhL0nqURobG3nwsQcZsNOAzheuwsLZC2tSztpYvnx5zbvGbYthL0nqcQbsNICh5w2tSVlN11S/p758+XJOPfVUpk2bxp577smkSZOYMWMGF1xwAQsWLGDbbbdl4sSJ7LDDDlx//fVcd911LF26lD322IObbrqJTTbZhFGjRrH11lszbdo0PvjBD/LpT3961WA6EcEDDzzAZpttVpPX1sxz9pIkVekvf/kLZ5xxBk8++SSbb74548ePZ8yYMdxxxx1MnTqV008/nW9+85sAHH/88Tz++OM88cQT7LPPPvz0pz9dVc6zzz7Lvffeyw9/+EN+8IMfMH78eBobG/njH/9I//79a95u9+wlSarS+973vlXD1p5yyilceeWVNDU1cdRRRwFFX/c77LADAE1NTVx66aXMmzePBQsWcMwxx6wq53Of+xx9+vQB4JBDDuGCCy7g5JNP5vjjj2fnnXeuebsNe0mSqhQR73m+2WabMWTIEB555JHVlh01ahS//OUv2XfffZk4cSL333//qnkth7i9+OKLOfbYY7nrrrs46KCDuPfee9l7771r2m4P40uSVKVZs2atCvbbbruNgw46iDlz5qyatmzZMp5++mkA5s+fzw477MCyZcu45ZZb2i3z+eefZ9iwYVx00UWMGDGCZ555pubtds9ektTjLJy9cI0urOusLHaqbtl99tmHG2+8kTPPPJMPfOADjBkzhmOOOYbzzjuPt99+m+XLlzN27FiGDBnC5ZdfzoEHHsiuu+7KsGHDmD9/fptljhs3jt///vf06dOHwYMH8/GPf7wmr6slw16S1KPUeohbdqquzEGDBjF9+vQ22/PAAw+sNv3ss8/m7LPPXm36xIkT3/P83//936tv61oy7CVJPcr6NMRtT+E5e0mSSs6wlySp5Ax7SVKPkJnd3YRus66v3bCXJK33+vXrx5tvvtkrAz8zefPNN+nXr99al+EFepKk9d7OO+/Myy+/zJw5c7q7Kd2iX79+69SzXt3DPiJGA18DdgCeBsZm5h+rWO9Q4H7gmcyszWgHkqQeqW/fvrz//e/v7mb0WHU9jB8RJwJXA1cC+wEPA3dHxC6drLcVMAm4r57tkySpN6j3OfsLgImZeX1mzsjMMcCrwOq9DLzXT4EbgdU7G5YkSWukbmEfERsBHwLuaTXrHuDDHaw3GtgeuKJebZMkqTep5579tkAf4PVW01+nCPPVRMQw4NvAyZm5orMKIuKMiJgSEVN660UbkiR1pituvWt9n0S0MY2I2Bi4HbgwM1+squDM6zJzRGaOGDhw4Lq3VJKkEqrn1fhzgRWsvhe/Havv7UNxtf5g4IaIuKEybQMgImI58InMbH1KQJIkdaJue/aZuRSYChzVatZRFFfltzYbGAY0tHhMAJ6r/N3WOpIkqRP1vs/+KuCmiHgMeAg4C9iRIsSJiEkAmfnFzFwGvGdw4oh4A3g3M2szaLEkSb1QXcM+MydHxDbApRSH6ZsoDse/VFmkw/vtJUnSuqt7D3qZeS1wbTvzRnay7mXAZTVvlCRJvYgD4UiSVHKGvSRJJWfYS5JUcoa9JEkl53j2krrM2LFjaWxs7PJ6GxoaGDduXJfXK60vDHtJXaaxsZEHH32c/gO77q7bxXNmdVld0vrKsJfUpfoP3IU9T7qky+p79vbvdFld0vrKc/aSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZzj2UuSOjV27FgaGxu7pe6GhgbGjRvXLXWXhWEvSepUY2MjDz76OP0H7tKl9S6eM6tL6ysrw16SVJX+A3dhz5Mu6dI6n739O11aX1l5zl6SpJIz7CVJKjnDXpKkkjPsJUkqOcNekqSSM+wlSSo5w16SpJIz7CVJKjnDXpKkkrMHvS7y7rzXaWx8lZEjR3Z53fYrve7cfj2X265nc/vVhmHfRVYsXUJu8C7vvvR4l9Y7/bUlXVpfWbn9ei63Xc/m9qsNw74LDd6+H5NP261L6zzxhhe6tL4yc/v1XG67ns3tt+48Zy9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnN3lSpLUysw3l7L4rcbSDMBj2EuS1MrCpStZtMECps2e1rX1zl5Yl3INe0mS2jBgpwEMPW9ol9bZdE1TXcr1nL0kSSVn2EuSVHJ1D/uIGB0RL0bEkoiYGhGHdbDsRyLi4Yh4MyIWR8QzEXFhvdsoSVKZ1fWcfUScCFwNjAYerPx7d0QMzsxZbayyALgGeApYBBwC/EdELMrMa+vZVkmSyqree/YXABMz8/rMnJGZY4BXgbPbWjgzp2bm7Zn5dGa+mJk3A78F2j0aIEmSOla3sI+IjYAPAfe0mnUP8OEqy9ivsuwfats6SZJ6j3ru2W8L9AFebzX9dWD7jlaMiJcj4l1gCnBtZk5oZ7kzImJKREyZM2dOLdosSVLpdMXV+NnqebQxrbXDgBHAWcDYiPhCmwVnXpeZIzJzxMCBA9e9pZIklVBVYR8Re0bEfRHRVHk+PCIu7WS1ucAKVt+L347V9/bfo3K+/qnMvB64CrismnZKkqTVVbtnfz1wCbAMIDOfBE7qaIXMXApMBY5qNeso4OE1bOPGa7C8JElqodpb7zbJzMciouW05VWsdxVwU0Q8BjxEcVh+R2ACQERMAsjML1aejwFeBP5SWf9w4ELA2+4kSVpL1Yb93IjYncq59oj4LMUtdB3KzMkRsQ1wKbAD0AR8IjNfqiyyS6tV+gDfAwZR/Jh4HriYyo8DSZK05qoN+3OA64C9I2I2xd73KdWsWOkMp80988wc2er5OKC24/pJktTLVRX2mfkC8I8RMQDYIDPn17dZkiSpVqoK+4i4Evh+Zs6rPN8K+JfM7OyKfEnqlWa+uZTFbzUycuTILq+7oaGBceM8SKq/q/Yw/scz8xvNTzLzrYj4BMW5eElSKwuXrmTRBguYNnta19Y7e2GX1qeeodqw7xMRG2fmuwAR0R9vh5OkDg3YaQBDzxvapXU2XdPUpfWpZ6g27G8G7ouIGyiuyD8duLFurZIkSTVT7QV634+Ip4AjKbq7vTwzf1vXlkmSpJqoejz7zLwbuLuObZEkSXVQbd/4x0fE/0bE2xHxTkTMj4h36t04SZK07qrds/8+8KnMnFHPxkiSpNqrdiCc1w16SZJ6pmr37KdExGTgl8C7zRMz8xd1aZUkSaqZasN+c2ARcHSLaQkY9pIkreeqvfXutHo3RJIk1Ue1V+PvGRH3RURT5fnwiLCrXEmSeoBqL9C7HrgEWAaQmU8CJ9WrUZIkqXaqDftNMvOxVtOW17oxkiSp9qoN+7kRsTvFRXlExGeBV+vWKkmSVDPVXo1/DnAdsHdEzAZeBE6pW6skSVLNVHs1/gvAP0bEAGCDzJxf32ZJkqRa6TDsI+KCdqYDkJlX1aFNkiSphjrbs9+sS1ohSZLqpsOwz8x/66qGSJKk+qjqnH1E9AO+BAwB+jVPz8zT69QuSZJUI9XeencTsD1wDPAHYGfAi/QkSeoBqg37PTLzX4GFmXkjcCwwrH7NkiRJtVJt2C+r/DsvIoYCWwCD6tIiSZJUU9V2qnNdRGwF/CtwJ7Bp5W9JkrSeqzbsb8jMFRTn63erY3skSVKNVXsY/8WIuC4ijozmHnUkSVKPUG3Y7wXcS9FH/syI+HFEHFq/ZkmSpFqpKuwzc3Fm/iwzjwcagM0pDulLkqT1XLV79kTERyLiWuDPFB3r/HPdWiVJkmqm2h70XgQagZ8BX8vMhXVtlSRJqplqr8bfNzPfqWtLJElSXVR7GH/7iLgvIpoAImJ4RFxax3ZJkqQaqTbsrwcuodKTXmY+CZxUr0ZJkqTaqTbsN8nMx1pNW17rxkiSpNqrNuznRsTuQAJExGeBV+vWKkmSVDPVXqB3DnAdsHdEzAZeBE6uW6skSVLNdBr2EbEBMCIz/zEiBgAbZKZj2UuS1EN0ehg/M1cC51b+XmjQS5LUs1R7zv53EXFhRLwvIrZuftS1ZZIkqSaqPWd/OsXFeaNbTXe4W6mOZr65lMVvNTJy5MgurbehoYFx48Z1aZ2S6qfasB9MEfSHUoT+H4EJ9WqUpMLCpStZtMECps2e1nV1zrY3bKlsqg37G4F3gGsqzz9fmeZgOFKdDdhpAEPPG9pl9TVd09RldUnqGtWG/V6ZuW+L57+PiCfq0SBJklRb1V6gNy0iDmp+EhEHAg/Vp0mSJKmWqt2zPxD4YkTMqjzfBZgREU8BmZnD69I6SZK0zqoN+4/VtRWSJKluqgr7zHyp3g2RJEn1Ue05e0mS1EMZ9pIklZxhL0lSydU97CNidES8GBFLImJqRBzWwbLHR8Q9ETEnIuZHxJ8i4tP1bqMkSWVW17CPiBOBq4Ergf2Ah4G7I2KXdlb5CPA/wLGV5e8C/rujHwiSJKlj1d56t7YuACZm5vWV52Mi4mPA2cAlrRfOzK+2mvRvEXEscBxFf/ySJGkN1W3PPiI2Aj4E3NNq1j3Ah9egqM2At2rVLkmSept6HsbfFugDvN5q+uvA9tUUEBHnADsDN7Uz/4yImBIRU+bMmbMubZUkqbS64mr8bPU82pi2mog4Afj/gJPb69QnM6/LzBGZOWLgwIHr3lJJkkqonmE/F1jB6nvx27H63v57VIL+JuCLmXlnfZonSVLvULewz8ylwFTgqFazjqK4Kr9NEfHPwM3AqMy8o17tkySpt6j31fhXATdFxGMUQ+KeBewITACIiEkAmfnFyvOTKPboLwQeiIjmowJLM/NvdW6rJEmlVNewz8zJEbENcCmwA9AEfKLFOfjW99ufVWnTuMqj2R+AkfVsqyRJZVXvPXsy81rg2nbmjezouSRJWnf2jS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklVzdO9VR95r55lIWv9XIyJEju7zuhoYGxo0b1/mCkqS6MuxLbuHSlSzaYAHTZk/r2npnL+zS+iRJ7TPse4EBOw1g6HlDu7TOpmuaurQ+SVL7PGcvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSydU97CNidES8GBFLImKOxLdaAAANqElEQVRqRBzWwbI7RMStEfFMRKyIiIn1bp8kSWVX17CPiBOBq4Ergf2Ah4G7I2KXdlbZGJgLfBf4Uz3bJklSb1HvPfsLgImZeX1mzsjMMcCrwNltLZyZMzPzvMycCPytzm2TJKlXqFvYR8RGwIeAe1rNugf4cL3qlSRJ71XPPfttgT7A662mvw5sX4sKIuKMiJgSEVPmzJlTiyIlSSqdrrgaP1s9jzamrV3Bmddl5ojMHDFw4MBaFClJUunUM+znAitYfS9+O1bf25ckSXVSt7DPzKXAVOCoVrOOorgqX5IkdYEN61z+VcBNEfEY8BBwFrAjMAEgIiYBZOYXm1eIiIbKn5sDKyvPl2bm9Dq3VZKkUqpr2Gfm5IjYBrgU2AFoAj6RmS9VFmnrfvtprZ5/CngJGFSvdkqSVGb13rMnM68Frm1n3sg2pkW92yRJUm9i3/iSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSyRn2kiSVnGEvSVLJGfaSJJWcYS9JUskZ9pIklZxhL0lSydU97CNidES8GBFLImJqRBzWyfIfqSy3JCJeiIiz6t1GSZLKrK5hHxEnAlcDVwL7AQ8Dd0fELu0s/37grspy+wHfAf49Ik6oZzslSSqzeu/ZXwBMzMzrM3NGZo4BXgXObmf5s4BXMnNMZfnrgRuBC+vcTkmSSmvDehUcERsBHwJ+0GrWPcCH21nt4Mr8ln4LnBoRfTNzWS3atnjOLJ69/Tu1KKpqK5e9y/TX4MQbXujSehctXQmzF9J0TVOX1rtw9kLYqT5lu/3qq57bDrp++/WmbQf+36uVsm2/yMzalwpExI7AbOAjmflAi+nfAk7OzL3aWOdZ4ObM/D8tph0O/AHYMTNfbbX8GcAZlad7AX+p+QtZf2wLzO3uRmituf16Lrddz1b27bdrZg7sbKG67dm30PrXRLQxrbPl25pOZl4HXLf2Tes5ImJKZo7o7nZo7bj9ei63Xc/m9ivU85z9XGAFsH2r6dsBr7ezzmvtLL8ceLOmrZMkqZeoW9hn5lJgKnBUq1lHUVxt35ZHgH9sY/kptTpfL0lSb1Pvq/GvAkZFxJcjYp+IuBrYEZgAEBGTImJSi+UnADtHxLjK8l8GRrH6RX69Ua84XVFibr+ey23Xs7n9qOMFeqsqiBgNfB3YAWgCzm++YC8i7gfIzJEtlv8I8CNgCPAK8L3MnFDXRkqSVGJ1D3tJktS97BtfkqSS6/VhHxH3R0RWHgd1d3t6ooiY2OI9/Gx3t2dt9YTPQkSMatHGH3d3e9Ynbr/1Q0/YDl2p1fvRbbcA9vqwr7iB4pqCqc0TImKriLgpIt6uPG6KiC07KiQijo+I30bEnMqGHVlN5Wsz+E9EXB0RUyrrzKyynoiIyyLilYhYXPkQDulknSERcUelXRkRl7Wx2Fcp3r8yaOuz8M2IeCgiFkZEVee9IuLwiLgzImZX3rdRVazTr/LD6cmIWNZ8TUsrkyvte6SqV9P7tLX9Zrb4sm1+fLejQir/T1qv81on6wxqY52MiI+1WKy3bL+1+n/UzvvX4fdhNd+7rQK3+XF7J+V2+t1X5efkeOCAjurqCoZ9YVFmvtbq9r5bgQ8CHwc+Vvn7pk7KGUBxW+EF1VYcaz/4zwYU4wZM6mS5lr4O/AswBtgfeAP4XURs1sE6mwAzgUuBF9taIDPfzswOvwh7kLY+CxsDvwDGrUE5m1JckPpVYHGV6/QBlgA/Bn7d1gKZubjyXi9dg7b0Jm1tP4D/QxE+zY8rqijrL63WGVZlGz7War3/aZ7Ri7bfuvw/+grvff9u7GT5ar93m3+AND/O7GT5Tr/7Kjr8nGTm34A5ndRVd13Rg16PExH7UPyHPTQzH65MOxP4Y0TslZltdsubmTdVlt12DapbNfhP5fmMiDiQYvCfn7e3UvPyEXEhcHRnlUREAGOB72bmzyvTTqUI/P8H+I926nkceLyy/DeqfE2lkpnfAog1OEWRmXdR/IgjIiZWuc5Cis8DETEc6PBIktbI/LX4Qbp8LX/EvlmiH781swb/j+atyfu3Bt+7i9aw3Gq/+9b2c9Kl3LNv28HAAt7b+c9DwELaH8RnXepqa/CfERHRt4b1vJ+id8JVdWXmYuABav+apPXNhRHxZkQ0Vg4nb1TFOrtVTsO8GBG3R8RuVdb1i4h4o3LIusdew9KNro6IuRHxeEScFRG1yqmTKuU+HRE/6OSI5ppY289Jl3LPvm3bA3OyxX2JmZkR8Qard+dbi7rubTXtdYptsy3FkMC1qqe57NZ11XGMM6nbXQNMo+hy+wDguxQ/fr/cwTp/oujQ6xmKLrsvBR6OiCGZ2V7X3Qsojsg9RNHF96eByRFxambeXIPX0Rt8C/g9xXt5JPBDiu/Bak67dORW4CWKvluGUJwu3ZfVe3hdU2vzOekWhn372rqApLNBfGpVV7uD/9SpLjtbUGll5lUtnj4ZEe9QhPBF7X0hZ+bdLZ9HxKPAC8CpFD2DtrXOXIpwajalcmj564BhX4XMvLzF08aI6AN8k3UM+8qgac2eiogXgD9FxAcz88/rUO4af066i4fx2/YasF3lPDew6pz3QNofxGdd6uqKwX+azymtycBEUhn9qfLvHtWukJkLgKeBD6xFXWu6jv7uT8DmEfEPNS53CsVAbTXdNuvwOak7w75tj1BcTX1wi2kH8/erPmtdV1cM/vMiReCvOmwVEf2Aw6j9a5LWZw2Vf6s+RVb5v7L3mqzToq5anYrrjRoo7lCZV+Nyh1Hc/VLTbbMOn5O68zB+GzJzRkT8BviPiPgKxaHu/wB+1XwlfkTsBNwHXJKZ/12ZtjWwC3+/inqPiJgHvNbB1ZoTgHMjYlyljkMozgF9vnmBiDgXODcz924xbQ+KHyQ7AhtFRPMX2PTKiIOtX1NW6vhmRDwDPEtxfmkBxfms5nLvAx7LzEsqzzcCBldm9wO2r9S1IDOf6+h9LIuI2AXYGhhUed78Xj9X+SXf1jqb8vc9xw2AXSrr/S0zZ3VQ12BgI4rzlJs215WZjTV4Kb1ORBwMHERxHvhtiltOfwTc2cl2+AHwf4FZFEe//pXix367t4FV7m5ZRnF9wErgU8A5wEW1eC09XWf/jyLiUxRHHh+huF31CIpbJq/LzHc7KLfD792I2B04meLumLkU32c/pNhOD3VQbqfffWvzOek2mdmrH8D9wI/bmL41xXm2dyqPm4EtW8wfRHGue1SLaaMq01o/LuukDR8B/gy8S7EHflar+ZcVm2q1drdV16AO6olKWa9S/Fr+AzC01TIzgYltvM7Wj/vbKD+Bz3b3Nq3DZ2FiO+/ByA7KGtnOOhM7acPMttartq29+dHWe0LRP8ajFHuGiykupLoM2KSTsm6nuJhrKTCb4jbYwZ2scyowneKunXcoDhWfsiaftTI81vb/EcXtztOA+ZX38CmKfio27KS+Ue2Ue1ll/vsq33VvUnzHPgdcDWzdSbmdfvdV+zlpUdaI7touvX4gnCh6KWvKzHO7uy09XaVXrM9l5h3d3Za10ZM+Cz2prV2lJ70nPamta6rMr21tRcQgih25/TNzSne0wXP2hTMiYkFE7N/dDemJImJCRLR5OLsHWq8/CxFxcuW9Pqy727KecvutH9br7dCVIuJuiov2urcd7tnHTkD/ytO/ZgfnhtS2iNgO2Lzy9NUseoLrcXrCZ6HSEUjzlcnzsrjdS7j91hc9YTt0pfXl/ej1YS9JUtl5GF+SpJIz7CVJKjnDXlK7ohgLvsPRxKpZRlL3MuwlSSo5w14SABHxy4iYWhkC9IxW8wZFxDMRcWNEPBkRd0TEJi0WGRMRf46IpyJi78o6B0TEwxExrfLvXpXpQyLisSiGm30yIta7fsSlsjHsJTU7PTM/BIwAzouIbVrN34ui69LhFD3EjW4xb25mfhD4CcUwr1D0Vnd4Zu5HMXTplZXpZwFXZ2ZDpa6X6/JqJK1i2Etqdl5EPEHRvez7WH3krr9mZnNf4jcDh7aY94vKv1Op9H0ObAH8V0Q0UfRHP6Qy/RHgGxFxEbBrZi6u6auQtBrDXhIRMZJi9MWDM3Nfij7K+7VarHWnHC2fN3cUsoK/D7B1OfD7zBxKMShMP4DMvBX4NEVf9b+NiI/W6GVIaodhLwmKvfC3MnNR5Zz7QW0ss0tlFDkoRmV8sIoyZ1f+HtU8MSJ2A17IzGuAO4Hh69JwSZ0z7CUB/AbYMCKepNgjf7SNZWYAp1aW2Zri/HxHvg98JyIeohg7vNmJQFNENFKM/T1pXRsvqWN2lyupU5VRu35VOSQvqYdxz16SpJJzz16SpJJzz16SpJIz7CVJKjnDXpKkkjPsJUkqOcNekqSSM+wlSSq5/x/Hr559qCjJ2gAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 576x432 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# melted.to_csv('/projects/trans_scratch/validations/workspace/szong/Cervical/notebook/animals_long.txt', index=False)\n",
    "plt.figure(figsize = (8, 6))\n",
    "sns.barplot(x = 'alphas', y = 'prevalence', hue = 'species', data = melted,\n",
    "            edgecolor = 'k', linewidth = 1.5);\n",
    "plt.xticks(size = 14); \n",
    "plt.yticks(size = 14)\n",
    "plt.title('Expected Value');"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "minimal sufficient python apis: will go through them at some point\n",
    "Attributes\n",
    "columns\n",
    "dtypes\n",
    "index\n",
    "shape\n",
    "T\n",
    "values\n",
    "\n",
    "Aggregation Methods\n",
    "all\n",
    "any\n",
    "count\n",
    "describe\n",
    "idxmax\n",
    "idxmin\n",
    "max\n",
    "mean\n",
    "median\n",
    "min\n",
    "mode\n",
    "nunique\n",
    "sum\n",
    "std\n",
    "var\n",
    "\n",
    "Non-Aggretaion Statistical Methods\n",
    "abs\n",
    "clip\n",
    "corr\n",
    "cov\n",
    "cummax\n",
    "cummin\n",
    "cumprod\n",
    "cumsum\n",
    "diff\n",
    "nlargest\n",
    "nsmallest\n",
    "pct_change\n",
    "prod\n",
    "quantile\n",
    "rank\n",
    "round\n",
    "\n",
    "Subset Selection\n",
    "head\n",
    "iloc\n",
    "loc\n",
    "tail\n",
    "\n",
    "Missing Value Handling\n",
    "dropna\n",
    "fillna\n",
    "interpolate\n",
    "isna\n",
    "notna\n",
    "\n",
    "Grouping\n",
    "expanding\n",
    "groupby\n",
    "pivot_table\n",
    "resample\n",
    "rolling\n",
    "\n",
    "Joining Data\n",
    "append\n",
    "merge\n",
    "\n",
    "Other\n",
    "asfreq\n",
    "astype\n",
    "copy\n",
    "drop\n",
    "drop_duplicates\n",
    "equals\n",
    "isin\n",
    "melt\n",
    "plot\n",
    "rename\n",
    "replace\n",
    "reset_index\n",
    "sample\n",
    "select_dtypes\n",
    "shift\n",
    "sort_index\n",
    "sort_values\n",
    "to_csv\n",
    "to_json\n",
    "to_sql\n",
    "Functions\n",
    "pd.concat\n",
    "pd.crosstab\n",
    "pd.cut\n",
    "pd.qcut\n",
    "pd.read_csv\n",
    "pd.read_json\n",
    "pd.read_sql\n",
    "pd.to_datetime\n",
    "pd.to_timedelta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### removed duplicate index, this works for multiple index as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>gene</th>\n",
       "      <th>symb</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>nd</th>\n",
       "      <th>cds_start</th>\n",
       "      <th>cds_end</th>\n",
       "      <th>status</th>\n",
       "      <th>chrn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_000014</td>\n",
       "      <td>alpha-2-macroglobulin precursor</td>\n",
       "      <td>A2M</td>\n",
       "      <td>2</td>\n",
       "      <td>chr12</td>\n",
       "      <td>0</td>\n",
       "      <td>9220303</td>\n",
       "      <td>9268558</td>\n",
       "      <td>9220418</td>\n",
       "      <td>9268445</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_000015</td>\n",
       "      <td>arylamine N-acetyltransferase 2</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>10</td>\n",
       "      <td>chr8</td>\n",
       "      <td>1</td>\n",
       "      <td>18248754</td>\n",
       "      <td>18258723</td>\n",
       "      <td>18257513</td>\n",
       "      <td>18258386</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      refseq                             gene  symb  locus_id    chr  strand  \\\n",
       "0  NM_000014  alpha-2-macroglobulin precursor   A2M         2  chr12       0   \n",
       "1  NM_000015  arylamine N-acetyltransferase 2  NAT2        10   chr8       1   \n",
       "\n",
       "      start        nd  cds_start   cds_end    status  chrn  \n",
       "0   9220303   9268558    9220418   9268445  Reviewed    12  \n",
       "1  18248754  18258723   18257513  18258386  Reviewed     8  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(40430, 12)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# use gistic refgenefile including both the cytoband and ref genes\n",
    "f1 = '/projects/da_workspace/software/gistic/rg.txt'\n",
    "df1 = pd.read_csv(f1)\n",
    "df1.head(2)\n",
    "# df1 = df1.set_index(['chrn', 'start', 'nd'])\n",
    "df1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([], dtype='int64')"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s = pd.Series(df1.head(700).index.duplicated())\n",
    "s[s].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>gene</th>\n",
       "      <th>symb</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>nd</th>\n",
       "      <th>cds_start</th>\n",
       "      <th>cds_end</th>\n",
       "      <th>status</th>\n",
       "      <th>chrn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>NM_000643</td>\n",
       "      <td>glycogen debranching enzyme isoform 1</td>\n",
       "      <td>AGL</td>\n",
       "      <td>178</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>100316044</td>\n",
       "      <td>100389579</td>\n",
       "      <td>100316598</td>\n",
       "      <td>100387207</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>608</th>\n",
       "      <td>NM_000644</td>\n",
       "      <td>glycogen debranching enzyme isoform 1</td>\n",
       "      <td>AGL</td>\n",
       "      <td>178</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>100316044</td>\n",
       "      <td>100389579</td>\n",
       "      <td>100316598</td>\n",
       "      <td>100387207</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>609</th>\n",
       "      <td>NM_000645</td>\n",
       "      <td>glycogen debranching enzyme isoform 2</td>\n",
       "      <td>AGL</td>\n",
       "      <td>178</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>100326765</td>\n",
       "      <td>100389579</td>\n",
       "      <td>100327027</td>\n",
       "      <td>100387207</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>610</th>\n",
       "      <td>NM_000646</td>\n",
       "      <td>glycogen debranching enzyme isoform 3</td>\n",
       "      <td>AGL</td>\n",
       "      <td>178</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>100316530</td>\n",
       "      <td>100389579</td>\n",
       "      <td>100318225</td>\n",
       "      <td>100387207</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>611</th>\n",
       "      <td>NM_000651</td>\n",
       "      <td>complement receptor type 1 isoform S precursor</td>\n",
       "      <td>CR1</td>\n",
       "      <td>1378</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>207669472</td>\n",
       "      <td>207815110</td>\n",
       "      <td>207669612</td>\n",
       "      <td>207812754</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612</th>\n",
       "      <td>NM_000655</td>\n",
       "      <td>L-selectin precursor</td>\n",
       "      <td>SELL</td>\n",
       "      <td>6402</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>169659805</td>\n",
       "      <td>169680843</td>\n",
       "      <td>169660924</td>\n",
       "      <td>169680678</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613</th>\n",
       "      <td>NM_000657</td>\n",
       "      <td>apoptosis regulator Bcl-2 beta isoform</td>\n",
       "      <td>BCL2</td>\n",
       "      <td>596</td>\n",
       "      <td>chr18</td>\n",
       "      <td>0</td>\n",
       "      <td>60985186</td>\n",
       "      <td>60986613</td>\n",
       "      <td>60985281</td>\n",
       "      <td>60985899</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>NM_000658</td>\n",
       "      <td>autoimmune regulator isoform 2</td>\n",
       "      <td>AIRE</td>\n",
       "      <td>326</td>\n",
       "      <td>chr21</td>\n",
       "      <td>1</td>\n",
       "      <td>45710273</td>\n",
       "      <td>45718110</td>\n",
       "      <td>45710689</td>\n",
       "      <td>45717610</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        refseq                                            gene  symb  \\\n",
       "607  NM_000643           glycogen debranching enzyme isoform 1   AGL   \n",
       "608  NM_000644           glycogen debranching enzyme isoform 1   AGL   \n",
       "609  NM_000645           glycogen debranching enzyme isoform 2   AGL   \n",
       "610  NM_000646           glycogen debranching enzyme isoform 3   AGL   \n",
       "611  NM_000651  complement receptor type 1 isoform S precursor   CR1   \n",
       "612  NM_000655                            L-selectin precursor  SELL   \n",
       "613  NM_000657          apoptosis regulator Bcl-2 beta isoform  BCL2   \n",
       "614  NM_000658                  autoimmune regulator isoform 2  AIRE   \n",
       "\n",
       "     locus_id    chr  strand      start         nd  cds_start    cds_end  \\\n",
       "607       178   chr1       1  100316044  100389579  100316598  100387207   \n",
       "608       178   chr1       1  100316044  100389579  100316598  100387207   \n",
       "609       178   chr1       1  100326765  100389579  100327027  100387207   \n",
       "610       178   chr1       1  100316530  100389579  100318225  100387207   \n",
       "611      1378   chr1       1  207669472  207815110  207669612  207812754   \n",
       "612      6402   chr1       0  169659805  169680843  169660924  169680678   \n",
       "613       596  chr18       0   60985186   60986613   60985281   60985899   \n",
       "614       326  chr21       1   45710273   45718110   45710689   45717610   \n",
       "\n",
       "       status  chrn  \n",
       "607  Reviewed     1  \n",
       "608  Reviewed     1  \n",
       "609  Reviewed     1  \n",
       "610  Reviewed     1  \n",
       "611  Reviewed     1  \n",
       "612  Reviewed     1  \n",
       "613  Reviewed    18  \n",
       "614  Reviewed    21  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>gene</th>\n",
       "      <th>symb</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>nd</th>\n",
       "      <th>cds_start</th>\n",
       "      <th>cds_end</th>\n",
       "      <th>status</th>\n",
       "      <th>chrn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>612</th>\n",
       "      <td>NM_000655</td>\n",
       "      <td>L-selectin precursor</td>\n",
       "      <td>SELL</td>\n",
       "      <td>6402</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>169659805</td>\n",
       "      <td>169680843</td>\n",
       "      <td>169660924</td>\n",
       "      <td>169680678</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        refseq                  gene  symb  locus_id   chr  strand      start  \\\n",
       "612  NM_000655  L-selectin precursor  SELL      6402  chr1       0  169659805   \n",
       "\n",
       "            nd  cds_start    cds_end    status  chrn  \n",
       "612  169680843  169660924  169680678  Reviewed     1  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.iloc[607:615]\n",
    "df1[df1.nd == 169680843]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>gene</th>\n",
       "      <th>symb</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>nd</th>\n",
       "      <th>cds_start</th>\n",
       "      <th>cds_end</th>\n",
       "      <th>status</th>\n",
       "      <th>chrn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_000014</td>\n",
       "      <td>alpha-2-macroglobulin precursor</td>\n",
       "      <td>A2M</td>\n",
       "      <td>2</td>\n",
       "      <td>chr12</td>\n",
       "      <td>0</td>\n",
       "      <td>9220303</td>\n",
       "      <td>9268558</td>\n",
       "      <td>9220418</td>\n",
       "      <td>9268445</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_000015</td>\n",
       "      <td>arylamine N-acetyltransferase 2</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>10</td>\n",
       "      <td>chr8</td>\n",
       "      <td>1</td>\n",
       "      <td>18248754</td>\n",
       "      <td>18258723</td>\n",
       "      <td>18257513</td>\n",
       "      <td>18258386</td>\n",
       "      <td>Reviewed</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      refseq                             gene  symb  locus_id    chr  strand  \\\n",
       "0  NM_000014  alpha-2-macroglobulin precursor   A2M         2  chr12       0   \n",
       "1  NM_000015  arylamine N-acetyltransferase 2  NAT2        10   chr8       1   \n",
       "\n",
       "      start        nd  cds_start   cds_end    status  chrn  \n",
       "0   9220303   9268558    9220418   9268445  Reviewed    12  \n",
       "1  18248754  18258723   18257513  18258386  Reviewed     8  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(40430, 9)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(29556, 9)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.loc[~df1.index.duplicated(keep='first')].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### use pd.cut to bin continuous values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    1\n",
       "1    1\n",
       "2    1\n",
       "Name: petal_length, dtype: int64"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# binning a column with continuous values into a categorical value\n",
    "import seaborn as sns\n",
    "import time\n",
    "import pandas as pd\n",
    "\n",
    "data = sns.load_dataset('iris')\n",
    "\n",
    "class_list = pd.cut(x=data.petal_length,\n",
    "                   bins=[0, 2, 5, 100],\n",
    "                   include_lowest=True,\n",
    "                   labels=[1, 2, 3]).astype(int)\n",
    "class_list[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### seaborn cmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAASwAAABLCAYAAADK+7ojAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAAfdJREFUeJzt27FqlEEYheFvwiraJZBgEcHGUtIYvAwru/R7N7kBCwvvxS52InZ6AWsnCGkmjY2Nv0sYZs+f5+kCU5wPwstuIK33XgAJjmYPAPhfggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYm6UHrbVtVW2rqurRk9ft5Hz0pmmenzydPWGo48dt9oRhNr9+zp4w1O/vu9kThvpWt7ve+9nSu7bPv+YcPXvZN1fX9xp2yK7fXcyeMNTb8/V+oD799HH2hKG+Xn2YPWGoN/Xjc+/9cunden+DgdURLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuIIVhADMECYggWEEOwgBiCBcQQLCCGYAExBAuI0Xrv/37Q2raqtn9+fFVVX0aPmui0qnazRwyy5tuq3JfuRe/9bOnRYrD+etzaTe/98l6zDtia71vzbVXueyh8JQRiCBYQY99gvR+y4nCs+b4131blvgdhr79hAczkKyEQQ7CAGIIFxBAsIIZgATHuABfGR4dfVLe1AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 360x72 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# you can just give a list of colors\n",
    "cmap = ['#0571b0', '#92c5de','#e0e0e0', '#f4a582', '#ca0020']\n",
    "sns.palplot(cmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hiv_pos_mfreq</th>\n",
       "      <th>hiv_neg_mfreq</th>\n",
       "      <th>mfreq</th>\n",
       "      <th>HTMCP-03-06-02260</th>\n",
       "      <th>HTMCP-03-06-02428</th>\n",
       "      <th>HTMCP-03-06-02139</th>\n",
       "      <th>HTMCP-03-06-02240</th>\n",
       "      <th>HTMCP-03-06-02262</th>\n",
       "      <th>HTMCP-03-06-02001</th>\n",
       "      <th>HTMCP-03-06-02411</th>\n",
       "      <th>...</th>\n",
       "      <th>HTMCP-03-06-02326</th>\n",
       "      <th>HTMCP-03-06-02185</th>\n",
       "      <th>HTMCP-03-06-02222</th>\n",
       "      <th>HTMCP-03-06-02197</th>\n",
       "      <th>HTMCP-03-06-02194</th>\n",
       "      <th>HTMCP-03-06-02332</th>\n",
       "      <th>HTMCP-03-06-02156</th>\n",
       "      <th>HTMCP-03-06-02046</th>\n",
       "      <th>HTMCP-03-06-02008</th>\n",
       "      <th>HTMCP-03-06-02037</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HIV_status</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>102</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Positive</td>\n",
       "      <td>...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Putative_histology</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>102</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Adeno</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>...</td>\n",
       "      <td>Adeno</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Squamous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows  121 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    hiv_pos_mfreq  hiv_neg_mfreq  mfreq HTMCP-03-06-02260  \\\n",
       "gene                                                                        \n",
       "HIV_status                    100            100    102          Positive   \n",
       "Putative_histology            100            100    102          Squamous   \n",
       "\n",
       "                   HTMCP-03-06-02428 HTMCP-03-06-02139 HTMCP-03-06-02240  \\\n",
       "gene                                                                       \n",
       "HIV_status                  Positive          Positive          Positive   \n",
       "Putative_histology          Squamous             Adeno          Squamous   \n",
       "\n",
       "                   HTMCP-03-06-02262 HTMCP-03-06-02001 HTMCP-03-06-02411  ...  \\\n",
       "gene                                                                      ...   \n",
       "HIV_status                  Positive          Positive          Positive  ...   \n",
       "Putative_histology          Squamous          Squamous          Squamous  ...   \n",
       "\n",
       "                   HTMCP-03-06-02326 HTMCP-03-06-02185 HTMCP-03-06-02222  \\\n",
       "gene                                                                       \n",
       "HIV_status                  Negative          Negative          Negative   \n",
       "Putative_histology             Adeno          Squamous          Squamous   \n",
       "\n",
       "                   HTMCP-03-06-02197 HTMCP-03-06-02194 HTMCP-03-06-02332  \\\n",
       "gene                                                                       \n",
       "HIV_status                  Negative          Negative          Negative   \n",
       "Putative_histology          Squamous          Squamous          Squamous   \n",
       "\n",
       "                   HTMCP-03-06-02156 HTMCP-03-06-02046 HTMCP-03-06-02008  \\\n",
       "gene                                                                       \n",
       "HIV_status                  Negative          Negative          Negative   \n",
       "Putative_histology          Squamous          Squamous          Squamous   \n",
       "\n",
       "                   HTMCP-03-06-02037  \n",
       "gene                                  \n",
       "HIV_status                  Negative  \n",
       "Putative_histology          Squamous  \n",
       "\n",
       "[2 rows x 121 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "hiv_pos_mfreq    False\n",
       "hiv_neg_mfreq    False\n",
       "mfreq            False\n",
       "Name: Putative_histology, dtype: bool"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "Index(['HTMCP-03-06-02110', 'HTMCP-03-06-02216', 'HTMCP-03-06-02203',\n",
       "       'HTMCP-03-06-02176'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HTMCP-03-06-02110</th>\n",
       "      <th>HTMCP-03-06-02216</th>\n",
       "      <th>HTMCP-03-06-02203</th>\n",
       "      <th>HTMCP-03-06-02176</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HIV_status</th>\n",
       "      <td>Positive</td>\n",
       "      <td>Positive</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Putative_histology</th>\n",
       "      <td>Adenosquamous</td>\n",
       "      <td>Adenosquamous</td>\n",
       "      <td>Adenosquamous</td>\n",
       "      <td>Adenosquamous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PIK3CA(35%)</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Non-synonymous</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAT1(19%)</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MLL2(15%)</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   HTMCP-03-06-02110 HTMCP-03-06-02216 HTMCP-03-06-02203  \\\n",
       "gene                                                                       \n",
       "HIV_status                  Positive          Positive          Negative   \n",
       "Putative_histology     Adenosquamous     Adenosquamous     Adenosquamous   \n",
       "PIK3CA(35%)                      NaN               NaN    Non-synonymous   \n",
       "FAT1(19%)                        NaN               NaN               NaN   \n",
       "MLL2(15%)                        NaN               NaN               NaN   \n",
       "\n",
       "                   HTMCP-03-06-02176  \n",
       "gene                                  \n",
       "HIV_status                  Negative  \n",
       "Putative_histology     Adenosquamous  \n",
       "PIK3CA(35%)                      NaN  \n",
       "FAT1(19%)                        NaN  \n",
       "MLL2(15%)                        NaN  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select columns based on values in a row, use series selection or transpose the df first\n",
    "mutation_f = '/projects/trans_scratch/validations/workspace/szong/Cervical/mutsig2cv/118_patients_old_vcfs/smgs_reviewed_details_with_frequency.txt.20190104'\n",
    "df = pd.read_csv(mutation_f, index_col='gene')\n",
    "df.head(2)\n",
    "s = df.loc['Putative_histology',:] == 'Adenosquamous'\n",
    "s[:3]\n",
    "s[s].index\n",
    "df[s[s].index].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### re.search?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### understand class self"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Area of Rectangle: 19200 cm^2\n",
      "Cost of rectangular field: Rs. 38400000 \n"
     ]
    }
   ],
   "source": [
    "class Rectangle:\n",
    "   def __init__(self, length, breadth, unit_cost=0):\n",
    "       self.length = length\n",
    "       self.breadth = breadth\n",
    "       self.unit_cost = unit_cost\n",
    "   \n",
    "   def get_perimeter(self):\n",
    "       return 2 * (self.length + self.breadth)\n",
    "   \n",
    "   def get_area(self):\n",
    "       return self.length * self.breadth\n",
    "   \n",
    "   def calculate_cost(self):\n",
    "       area = self.get_area()\n",
    "       return area * self.unit_cost\n",
    "# breadth = 120 cm, length = 160 cm, 1 cm^2 = Rs 2000\n",
    "r = Rectangle(160, 120, 2000)\n",
    "print(\"Area of Rectangle: %s cm^2\" % (r.get_area()))\n",
    "print(\"Cost of rectangular field: Rs. %s \" %(r.calculate_cost()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "to understand python object and self\n",
    "https://micropyramid.com/blog/understand-self-and-__init__-method-in-python-class/\n",
    "\n",
    "* class: a set of things (objects) having some properties in common and are different from others by type, kind, qulity etc. class is the blueprint for individual object with exactly the same behaviours.\n",
    "* object: one of the instances (examples) of the class, which can perform the functionalities defined in the class.\n",
    "* self: represent the instance of the class. by using self, we can access the attributes and methods of the class. \n",
    "* __init__: is the contructor, we call this reserved method to create an object from a class. It allows the class to initialize the attributes of a class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "string = 'ABCDCEDCXBGECDCCDXA'\n",
    "sub_string = 'CDC'\n",
    "def count_substring(string, sub_string):\n",
    "   counts = 0\n",
    "   for idx in range(len(string)-2):\n",
    "        part = string[idx:idx+3]\n",
    "        if part == sub_string:\n",
    "            counts += 1\n",
    "   return counts\n",
    "count_substring(string, sub_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ABCD'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "string[0:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: 'b'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-35-8a635b9ba9ae>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m__name__\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'__main__'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0;31m# so student_marks is a dictionary, key is the student name, values is a list of three marks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mstudent_marks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0m_\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: 'b'"
     ]
    }
   ],
   "source": [
    "if __name__ == '__main__':\n",
    "    n = int(input())\n",
    "    # so student_marks is a dictionary, key is the student name, values is a list of three marks\n",
    "    student_marks = {}\n",
    "    for _ in range(n):\n",
    "        name, *line = input().split()\n",
    "        scores = list(map(float, line))\n",
    "        student_marks[name] = scores\n",
    "    query_name = input()\n",
    "    marks = student_marks[query_name]\n",
    "    average_mark = sum(marks)/len(marks)\n",
    "    print(\"{0:.2f}\".format(average_mark))\n",
    "# main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5.00\n"
     ]
    }
   ],
   "source": [
    "print(\"{0:.2f}\".format(5.0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:.2f\n"
     ]
    }
   ],
   "source": [
    "print(\":.2f\".ormat(5.0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'iamKKKKKmartateverything'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "string = 'iamsmartateverything'\n",
    "position = 3\n",
    "character = 'KKKKK'\n",
    "def mutate_string(string, position, character):\n",
    "    l = list(string)\n",
    "    l[position] = character \n",
    "    s = ''.join(l) \n",
    "    return s\n",
    "mutate_string(string, position, character)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def show_this(astring):\n",
    "    print(s) # this is wrong because s comes from nowhere\n",
    "    print(astring) # this is correct because when you call function show_this you have to specify what astring is\n",
    "show_this('iamsmart')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.23606797749979"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "math.sqrt(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = ['a', 'b', 'c']\n",
    "b, *c = a\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'a'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "['b', 'c']"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b\n",
    "c"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ok this is the formatting you can do in notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "this is a **test** cell. you can add a [link](https://docs.python.org/3/tutorial/classes.html#a-first-look-at-classes) there. sometimes, you would like to use a *italics* format. i am not sure what `inline really` means.\n",
    ">Also, you can blockquote somethiend.ffffffffffffffffffff aaaaaaaaaaaaaaaa aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa dkdfjdfjd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50005 -100000000\n",
      "50005 33\n",
      "50005 33\n",
      "50005 33\n",
      "50005 33\n",
      "50005 50000\n",
      "50005 50000\n",
      "50005 50000\n",
      "55555 50005\n",
      "55555 50005\n",
      "55555 50005\n",
      "55555 50005\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "50005"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arr = [50005, 33, -551,2,3,50000, 50000, 4,55555, 55555, 5, 5]\n",
    "\n",
    "def second(arr):\n",
    "    largest = -100000000\n",
    "    second = -10000000\n",
    "    for idx in range(len(arr)):\n",
    "        num = arr[idx]\n",
    "        if num > largest:\n",
    "            second = largest\n",
    "            largest = num\n",
    "        elif num == largest:\n",
    "            pass # means do nothing\n",
    "        elif num > second:\n",
    "            second = num\n",
    "        print(largest, second)\n",
    "    return second\n",
    "second(arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[5, 4, 3, 2, 1]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(np.unique(n), reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n = [1,2,3,4,5, 5, 5]\n",
    "a = sorted(list(set(n)), reverse=True)\n",
    "a[1]\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
